<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>_systemPipeRdata_: Workflow templates and sample data • systemPipeRdata</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="_systemPipeRdata_: Workflow templates and sample data">
<meta property="og:description" content="systemPipeRdata">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeRdata</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/systemPipeRdata.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/systemPipeChIPseq.html">ChIP-Seq Workflow Template</a>
    </li>
    <li>
      <a href="../articles/systemPipeRIBOseq.html">RIBO-Seq Workflow Template</a>
    </li>
    <li>
      <a href="../articles/systemPipeRNAseq.html">RNA-Seq Workflow Template</a>
    </li>
    <li>
      <a href="../articles/systemPipeVARseq.html">VAR-Seq Workflow Template</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/tgirke/systemPipeRdata/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>
<em>systemPipeRdata</em>: Workflow templates and sample data</h1>
                        <h4 data-toc-skip class="author">Author: Daniela Cassol (<a href="mailto:danielac@ucr.edu" class="email">danielac@ucr.edu</a>) and Thomas Girke (<a href="mailto:thomas.girke@ucr.edu" class="email">thomas.girke@ucr.edu</a>)</h4>
            
            <h4 data-toc-skip class="date">Last update: 27 April, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/tgirke/systemPipeRdata/blob/HEAD/vignettes/systemPipeRdata.Rmd" class="external-link"><code>vignettes/systemPipeRdata.Rmd</code></a></small>
      <div class="hidden name"><code>systemPipeRdata.Rmd</code></div>

    </div>

    
    
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<!---
- Compile from command-line
Rscript -e "rmarkdown::render('systemPipeRdata.Rmd', c('BiocStyle::html_document'), clean=F); knitr::knit('systemPipeRdata.Rmd', tangle=TRUE)"

--><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><p><strong>Note:</strong> the most recent version of this vignette can be found <a href="https://github.com/tgirke/systemPipeRdata/blob/master/vignettes/systemPipeRdata.Rmd" class="external-link">here</a>.</p>
<p><strong>Note:</strong> if you use <em><code>systemPipeR</code></em> and <em><code>systemPipeRdata</code></em> in published research, please cite:</p>
<p>Backman, T.W.H and Girke, T. (2016). <em>systemPipeR</em>: Workflow and Report Generation Environment. <em>BMC Bioinformatics</em>, 17: 388. <a href="https://doi.org/10.1186/s12859-016-1241-0" class="external-link">10.1186/s12859-016-1241-0</a>.</p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><a href="https://github.com/tgirke/systemPipeRdata" class="external-link"><em><code>systemPipeRdata</code></em></a> is a helper package to generate with a single command workflow templates that are intended to be used by its parent package <a href="http://www.bioconductor.org/packages/devel/bioc/html/systemPipeR.html" class="external-link"><em><code>systemPipeR</code></em></a> <span class="citation">(H Backman and Girke 2016)</span>. The <em>systemPipeR</em> project provides a suite of R/Bioconductor packages for designing, building and running end-to-end analysis workflows on local machines, HPC clusters and cloud systems, while generating at the same time publication quality analysis reports.</p>
<p>To test workflows quickly or design new ones from existing templates, users can generate with a single command workflow instances fully populated with sample data and parameter files required for running a chosen workflow. Pre-configured directory structure of the workflow environment and the sample data used by <em><code>systemPipeRdata</code></em> are described <a href="http://bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#load-sample-data-and-workflow-templates" class="external-link">here</a>.</p>
<p><em><code>systemPipeRdata</code></em> package provides a demo sample FASTQ files used in the workflow reporting vignettes. The chosen data set <a href="http://www.ncbi.nlm.nih.gov/sra/?term=SRP010938" class="external-link"><code>SRP010938</code></a> obtains 18 paired-end (PE) read sets from <em>Arabidposis thaliana</em> <span class="citation">(Howard et al. 2013)</span>. To minimize processing time during testing, each FASTQ file has been subsetted to 90,000-100,000 randomly sampled PE reads that map to the first 100,000 nucleotides of each chromosome of the <em>A. thalina</em> genome. The corresponding reference genome sequence (FASTA) and its GFF annotation files (provided in the same download) have been truncated accordingly. This way the entire test sample data set requires less than 200MB disk storage space. A PE read set has been chosen for this test data set for flexibility, because it can be used for testing both types of analysis routines requiring either SE (single-end) reads or PE reads.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>The <em><code>systemPipeRdata</code></em> package is available at <a href="http://www.bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html" class="external-link">Bioconductor</a> and can be installed from within R as follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"systemPipeRdata"</span><span class="op">)</span></code></pre></div>
<p>Also, it is possible to install the development version from <a href="http://www.bioconductor.org/packages/devel/data/experiment/html/systemPipeRdata.html" class="external-link">Bioconductor</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"systemPipeRdata"</span>, version <span class="op">=</span> <span class="st">"devel"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>,
    dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  <span class="co"># Installs Devel version from Bioconductor</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="loading-package-and-documentation">Loading package and documentation<a class="anchor" aria-label="anchor" href="#loading-package-and-documentation"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/tgirke/systemPipeRdata" class="external-link">"systemPipeRdata"</a></span><span class="op">)</span>  <span class="co"># Loads the package</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span>help <span class="op">=</span> <span class="st">"systemPipeRdata"</span><span class="op">)</span>  <span class="co"># Lists package info</span>
<span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span><span class="st">"systemPipeRdata"</span><span class="op">)</span>  <span class="co"># Opens vignette</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="starting-with-pre-configured-workflow-templates">Starting with pre-configured workflow templates<a class="anchor" aria-label="anchor" href="#starting-with-pre-configured-workflow-templates"></a>
</h2>
<p>Load one of the available workflows into your current working directory. The following does this for the <em><code>rnaseq</code></em> workflow template. The name of the resulting workflow directory can be specified under the <em><code>mydirname</code></em> argument. The default <em><code>NULL</code></em> uses the name of the chosen workflow. An error is issued if a directory of the same name and path exists already.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"systemPipeR/SPrnaseq"</span>, mydirname <span class="op">=</span> <span class="st">"rnaseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"rnaseq"</span><span class="op">)</span></code></pre></div>
<p>On Linux and OS X systems the same can be achieved from the command-line of a terminal with the following commands.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> Rscript <span class="at">-e</span> <span class="st">"systemPipeRdata::genWorkenvir(workflow='systemPipeR/SPrnaseq', mydirname='rnaseq')"</span></span></code></pre></div>
<div class="section level3">
<h3 id="build-run-and-visualize-the-workflow-template">Build, run and visualize the workflow template<a class="anchor" aria-label="anchor" href="#build-run-and-visualize-the-workflow-template"></a>
</h3>
<ul>
<li>Build workflow from RMarkdown file</li>
</ul>
<p>This template provides some common steps for a <code>RNAseq</code> workflow. One can add, remove, modify workflow steps by operating on the <code>sal</code> object.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html" class="external-link">SPRproject</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/importWF.html" class="external-link">importWF</a></span><span class="op">(</span><span class="va">sal</span>, file_path <span class="op">=</span> <span class="st">"systemPipeVARseq.Rmd"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<ul>
<li>Running workflow</li>
</ul>
<p>Next, we can run the entire workflow from R with one command:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html" class="external-link">runWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<ul>
<li>Visualize workflow</li>
</ul>
<p><em><code>systemPipeR</code></em> workflows instances can be visualized with the <code>plotWF</code> function.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/plotWF.html" class="external-link">plotWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<ul>
<li>Report generation</li>
</ul>
<p><em><code>systemPipeR</code></em> compiles all the workflow execution logs in one central location, making it easier to check any standard output (<code>stdout</code>) or standard error (<code>stderr</code>) for any command-line tools used on the workflow or the R code stdout.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderLogs.html" class="external-link">renderLogs</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<p>Also, the technical report can be generated using <code>renderReport</code> function.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderReport.html" class="external-link">renderReport</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="workflow-templates-collection">Workflow templates collection<a class="anchor" aria-label="anchor" href="#workflow-templates-collection"></a>
</h2>
<p>A collection of workflow templates are available, and it is possible to browse the current availability, as follows:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/availableWF.html">availableWF</a></span><span class="op">(</span>github <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>This function returns the list of workflow templates available within the package and <a href="https://github.com/systemPipeR" class="external-link">systemPipeR Organization</a> on GitHub. Each one listed template can be created as described above.</p>
<p>The workflow template choose from Github will be installed as an R package, and also it creates the environment with all the settings and files to run the demo analysis.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow<span class="op">=</span><span class="st">"systemPipeR/SPrnaseq"</span>, mydirname<span class="op">=</span><span class="st">"NULL"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"SPrnaseq"</span><span class="op">)</span></code></pre></div>
<p>Besides, it is possible to choose different versions of the workflow template, defined through other branches on the GitHub Repository. By default, the <em><code>master</code></em> branch is selected, however, it is possible to define a different branch with the <em><code>ref</code></em> argument.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow<span class="op">=</span><span class="st">"systemPipeR/SPrnaseq"</span>, ref <span class="op">=</span> <span class="st">"singleMachine"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"SPrnaseq"</span><span class="op">)</span></code></pre></div>
<div class="section level3">
<h3 id="download-a-specific-r-markdown-file">Download a specific R Markdown file<a class="anchor" aria-label="anchor" href="#download-a-specific-r-markdown-file"></a>
</h3>
<p>Also, it is possible to download a specific workflow script for your analysis. The URL can be specified under <em><code>url</code></em> argument and the R Markdown file name in the <em><code>urlname</code></em> argument. The default <em><code>NULL</code></em> copies the current version available in the chose template.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow<span class="op">=</span><span class="st">"systemPipeR/SPrnaseq"</span>, url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/systemPipeR/systemPipeRNAseq/cluster/vignettes/systemPipeRNAseq.Rmd"</span>, 
             urlname <span class="op">=</span> <span class="st">"rnaseq_V-cluster.Rmd"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"rnaseq"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="dynamic-generation-of-workflow-template">Dynamic generation of workflow template<a class="anchor" aria-label="anchor" href="#dynamic-generation-of-workflow-template"></a>
</h2>
<p>It is possible to create a new workflow structure from RStudio menu <code>File -&gt; New File -&gt; R Markdown -&gt; From Template -&gt; systemPipeR New WorkFlow</code>. This interactive option creates the same environment as demonstrated above.</p>
<p><img src="results/rstudio.png"><strong>Figure 1:</strong> Selecting workflow template within RStudio.</p>
</div>
<div class="section level2">
<h2 id="directory-structure">Directory Structure<a class="anchor" aria-label="anchor" href="#directory-structure"></a>
</h2>
<p>The workflow templates generated by <em><code>genWorkenvir</code></em> contain the following preconfigured directory structure:</p>
<ul>
<li>
<span style="color:green"><em><strong>workflow/</strong></em></span> (<em>e.g.</em> <em>rnaseq/</em>)
<ul>
<li>This is the root directory of the R session running the workflow.</li>
<li>Run script ( <em>*.Rmd</em>) and sample annotation (<em>targets.txt</em>) files are located here.</li>
<li>Note, this directory can have any name (<em>e.g.</em> <span style="color:green"><em><strong>rnaseq</strong></em></span>, <span style="color:green"><em><strong>varseq</strong></em></span>). Changing its name does not require any modifications in the run script(s).</li>
<li>
<strong>Important subdirectories</strong>:
<ul>
<li>
<span style="color:green"><em><strong>param/</strong></em></span>
<ul>
<li>Stores non-CWL parameter files such as: <em>*.param</em>, <em>*.tmpl</em> and <em>*.run.sh</em>. These files are only required for backwards compatibility to run old workflows using the previous custom command-line interface.</li>
<li>
<span style="color:green"><em><strong>param/cwl/</strong></em></span>: This subdirectory stores all the CWL parameter files. To organize workflows, each can have its own subdirectory, where all <code>CWL param</code> and <code>input.yml</code> files need to be in the same subdirectory.</li>
</ul>
</li>
<li>
<span style="color:green"><em><strong>data/</strong></em> </span>
<ul>
<li>FASTQ files</li>
<li>FASTA file of reference (<em>e.g.</em> reference genome)</li>
<li>Annotation files</li>
<li>etc.</li>
</ul>
</li>
<li>
<span style="color:green"><em><strong>results/</strong></em></span>
<ul>
<li>Analysis results are usually written to this directory, including: alignment, variant and peak files (BAM, VCF, BED); tabular result files; and image/plot files</li>
<li>Note, the user has the option to organize results files for a given sample and analysis step in a separate subdirectory.</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<p><strong>Note</strong>: Directory names are indicated in <span style="color:grey"><strong><em>green</em></strong></span>. Users can change this structure as needed, but need to adjust the code in their workflows accordingly.</p>
<center>
<img src="results/directory.png">
</center>
<p><strong>Figure 2:</strong> <em>systemPipeR’s</em> preconfigured directory structure.</p>
<div class="section level3">
<h3 id="return-paths-to-sample-data">Return paths to sample data<a class="anchor" aria-label="anchor" href="#return-paths-to-sample-data"></a>
</h3>
<p>The location of the sample data provided by <em><code>systemPipeRdata</code></em> can be returned as a <em><code>list</code></em>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/pathList.html">pathList</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## $targets</span>
<span class="co">## [1] "/tmp/Rtmp0LR5iI/temp_libpathb28204df172e1/systemPipeRdata/extdata/param/targets.txt"</span>
<span class="co">## </span>
<span class="co">## $targetsPE</span>
<span class="co">## [1] "/tmp/Rtmp0LR5iI/temp_libpathb28204df172e1/systemPipeRdata/extdata/param/targetsPE.txt"</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="version-information">Version information<a class="anchor" aria-label="anchor" href="#version-information"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R Under development (unstable) (2021-10-25 r81105)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.4 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span>
<span class="co">## LAPACK: /home/dcassol/src/R-devel/lib/libRlapack.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets </span>
<span class="co">## [6] methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">## [1] systemPipeRdata_2.1.1 BiocStyle_2.23.1     </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##  [1] formatR_1.12           XVector_0.35.0        </span>
<span class="co">##  [3] bslib_0.3.1            compiler_4.2.0        </span>
<span class="co">##  [5] BiocManager_1.30.17    jquerylib_0.1.4       </span>
<span class="co">##  [7] GenomeInfoDb_1.31.10   remotes_2.4.2         </span>
<span class="co">##  [9] zlibbioc_1.41.0        bitops_1.0-7          </span>
<span class="co">## [11] tools_4.2.0            digest_0.6.29         </span>
<span class="co">## [13] jsonlite_1.8.0         evaluate_0.15         </span>
<span class="co">## [15] memoise_2.0.1          rlang_1.0.2           </span>
<span class="co">## [17] cli_3.3.0              yaml_2.3.5            </span>
<span class="co">## [19] pkgdown_2.0.3          xfun_0.30             </span>
<span class="co">## [21] fastmap_1.1.0          GenomeInfoDbData_1.2.8</span>
<span class="co">## [23] stringr_1.4.0          knitr_1.38            </span>
<span class="co">## [25] desc_1.4.1             Biostrings_2.63.3     </span>
<span class="co">## [27] fs_1.5.2               sass_0.4.1            </span>
<span class="co">## [29] S4Vectors_0.33.17      systemfonts_1.0.4     </span>
<span class="co">## [31] IRanges_2.29.1         stats4_4.2.0          </span>
<span class="co">## [33] rprojroot_2.0.3        R6_2.5.1              </span>
<span class="co">## [35] textshaping_0.3.6      rmarkdown_2.14        </span>
<span class="co">## [37] bookdown_0.26          purrr_0.3.4           </span>
<span class="co">## [39] magrittr_2.0.3         codetools_0.2-18      </span>
<span class="co">## [41] htmltools_0.5.2        BiocGenerics_0.41.2   </span>
<span class="co">## [43] ragg_1.2.2             stringi_1.7.6         </span>
<span class="co">## [45] RCurl_1.98-1.6         cachem_1.0.6          </span>
<span class="co">## [47] crayon_1.5.1</span></code></pre>
</div>
<div class="section level2">
<h2 id="funding">Funding<a class="anchor" aria-label="anchor" href="#funding"></a>
</h2>
<p>This project was supported by funds from the National Institutes of Health (NIH) and the National Science Foundation (NSF).</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-H_Backman2016-bt" class="csl-entry">
H Backman, Tyler W, and Thomas Girke. 2016. <span>“<span class="nocase">systemPipeR: NGS workflow and report generation environment</span>.”</span> <em>BMC Bioinformatics</em> 17 (1): 388. <a href="https://doi.org/10.1186/s12859-016-1241-0" class="external-link">https://doi.org/10.1186/s12859-016-1241-0</a>.
</div>
<div id="ref-Howard2013-fq" class="csl-entry">
Howard, Brian E, Qiwen Hu, Ahmet Can Babaoglu, Manan Chandra, Monica Borghi, Xiaoping Tan, Luyan He, et al. 2013. <span>“High-Throughput <span>RNA</span> Sequencing of Pseudomonas-Infected Arabidopsis Reveals Hidden Transcriptome Complexity and Novel Splice Variants.”</span> <em>PLoS One</em> 8 (10): e74183. <a href="https://doi.org/10.1371/journal.pone.0074183" class="external-link">https://doi.org/10.1371/journal.pone.0074183</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="http://girke.bioinformatics.ucr.edu/" class="external-link">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
