<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>VAR-Seq Workflow Template • systemPipeRdata</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="VAR-Seq Workflow Template">
<meta property="og:description" content="systemPipeRdata">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeRdata</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/systemPipeRdata.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/systemPipeChIPseq.html">ChIP-Seq Workflow Template</a>
    </li>
    <li>
      <a href="../articles/systemPipeRIBOseq.html">RIBO-Seq Workflow Template</a>
    </li>
    <li>
      <a href="../articles/systemPipeRNAseq.html">RNA-Seq Workflow Template</a>
    </li>
    <li>
      <a href="../articles/systemPipeVARseq.html">VAR-Seq Workflow Template</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/tgirke/systemPipeRdata/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>VAR-Seq Workflow Template</h1>
                        <h4 data-toc-skip class="author">Author: Daniela Cassol (<a href="mailto:danielac@ucr.edu" class="email">danielac@ucr.edu</a>) and Thomas Girke (<a href="mailto:thomas.girke@ucr.edu" class="email">thomas.girke@ucr.edu</a>)</h4>
            
            <h4 data-toc-skip class="date">Last update: 27 April, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/tgirke/systemPipeRdata/blob/HEAD/vignettes/systemPipeVARseq.Rmd" class="external-link"><code>vignettes/systemPipeVARseq.Rmd</code></a></small>
      <div class="hidden name"><code>systemPipeVARseq.Rmd</code></div>

    </div>

    
    
<!--
# Compile from command-line
Rscript -e "rmarkdown::render('systemPipeVARseq.Rmd', c('BiocStyle::html_document'), clean=FALSE); knitr::knit('systemPipeVARseq.Rmd', tangle=TRUE)""
-->
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Users want to provide here background information about the design of their VAR-Seq project.</p>
<p>This report describes the analysis of a VAR-Seq project studying the genetic differences among several strains … from <em>organism</em> ….</p>
<div class="section level3">
<h3 id="experimental-design">Experimental design<a class="anchor" aria-label="anchor" href="#experimental-design"></a>
</h3>
<p>Typically, users want to specify here all information relevant for the analysis of their VAR-Seq study. This includes detailed descriptions of FASTQ files, experimental design, reference genome, gene annotations, etc.</p>
</div>
</div>
<div class="section level2">
<h2 id="workflow-environment">Workflow environment<a class="anchor" aria-label="anchor" href="#workflow-environment"></a>
</h2>
<div class="section level3">
<h3 id="generate-workflow-environment">Generate workflow environment<a class="anchor" aria-label="anchor" href="#generate-workflow-environment"></a>
</h3>
<p><a href="http://bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html" class="external-link"><em>systemPipeRdata</em></a> package is a helper package to generate a fully populated <a href="http://bioconductor.org/packages/release/bioc/html/systemPipeR.html" class="external-link"><em>systemPipeR</em></a> workflow environment in the current working directory with a single command. All the instruction for generating the pre-configured workflow templates are provide in the <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRdata.html#1_Introduction" class="external-link"><em>systemPipeRdata</em> vignette</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">systemPipeRdata</span><span class="fu">::</span><span class="fu"><a href="../reference/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"varseq"</span>, mydirname <span class="op">=</span> <span class="st">"varseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"varseq"</span><span class="op">)</span></code></pre></div>
<p>This step can be skipped if you already have the environment to run the analysis. If not, you can run it, and it will create the directory structure and populate all the necessary param and demo data files.</p>
<p>After building and loading the workflow environment generated by <code>genWorkenvir</code> from <code>systemPipeRdata</code> all data inputs are stored in a <code>data/</code> directory and all analysis results will be written to a separate <code>results/</code> directory, while the <code>systemPipeVARseq.Rmd</code> script and the <code>targets</code> file are expected to be located in the parent directory. The R session is expected to run from this parent directory. Additional parameter files are stored under <code>param/</code>.</p>
<p>To work with real data, users want to organize their own data similarly and substitute all test data for their own data. To rerun an established workflow on new data, the initial <code>targets</code> file along with the corresponding FASTQ files are usually the only inputs the user needs to provide.</p>
<p>For more details, please consult the documentation <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html" class="external-link">here</a>. More information about the <code>targets</code> files from <em>systemPipeR</em> can be found <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#42_Structure_of_initial_targets_data" class="external-link">here</a>.</p>
</div>
<div class="section level3">
<h3 id="build-the-workflow-with-a-single-command">Build the Workflow with a single command<a class="anchor" aria-label="anchor" href="#build-the-workflow-with-a-single-command"></a>
</h3>
<p>This template provides some common steps for a <code>VARseq</code> workflow. One can add, remove, modify workflow steps by operating on the <code>SYSargsList</code> workflow object. For more details of all the features and utilities, please consult the <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html" class="external-link">main vignette</a>.</p>
<p>To initiate a VARseq workflow, this entire Rmarkdown file will be imported as a <code>SYSargsList</code> workflow object, by using the <code>importWF("systemPipeVARseq.Rmd")</code> command.</p>
<p>In this template, code chunks with the option <code>spr = TRUE'</code> will be added to the workflow. Other R code chunks without this option will be ignored. The option <code>eval = FALSE</code> can be ignored when imported and build the workflow object. Please be aware of this possibility.</p>
<p>The template can provide more than one alternative for each step, such as different mapping methods, that will receive the <code>mandatory</code> or <code>optional</code> flag. One can run just the <code>mandatory</code> steps, <code>ALL</code>, or <code>optional</code> steps when running the workflow.</p>
<p>Also, each one of the steps can be run on compute clusters (<code>compute</code> option) or on the current session, here called <code>management</code> session. For the demonstration of this template, a <code>management</code> session will be chosen.</p>
</div>
<div class="section level3">
<h3 id="workflow-initialization">Workflow initialization<a class="anchor" aria-label="anchor" href="#workflow-initialization"></a>
</h3>
<p>The other alternative is to initialize the workflow and append each of the steps in the workflow object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html" class="external-link">SPRproject</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="required-packages-and-resources">Required packages and resources<a class="anchor" aria-label="anchor" href="#required-packages-and-resources"></a>
</h3>
<p><em><code>systemPipeR</code></em> workflows can be designed and built from start to finish with a single command, importing from an R Markdown file or stepwise in interactive mode from the R console. This tutorial will demonstrate how to build the workflow in an interactive mode, appending each step. The workflow is constructed by connecting each step via <code>appendStep</code> method. Each <code>SYSargsList</code> instance contains instructions needed for processing a set of input files with a specific command-line or R software and the paths to the corresponding outfiles generated by a particular tool/step.</p>
<p>The <code>systemPipeR</code> package needs to be loaded <span class="citation">(H Backman and Girke 2016)</span>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Some samples in the test dataset do not work well in</span>
<span class="co"># VARseq, and VARseq workflow takes long time to process</span>
<span class="co"># each sample. To better test and speed up the test</span>
<span class="co"># workflow, sample set is reduced to the first 8 samples.</span>
<span class="co"># Please REMOVE the next two lines in your real analysis</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu">crayon</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/crayon/man/crayon.html" class="external-link">red</a></span><span class="op">$</span><span class="fu">bold</span><span class="op">(</span><span class="st">"Some samples in targets are removed for test workflow. Please change the template to disable this in your real analysis.\n"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/writeLines.html" class="external-link">writeLines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="st">"targetsPE.txt"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">13</span><span class="op">]</span>, <span class="st">"targetsPE.txt"</span><span class="op">)</span>

<span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://systempipe.org/" class="external-link">systemPipeR</a></span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"load_SPR"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fastq-quality-report">FASTQ quality report<a class="anchor" aria-label="anchor" href="#fastq-quality-report"></a>
</h3>
<p>The following <code>seeFastq</code> and <code>seeFastqPlot</code> functions generate and plot a series of useful quality statistics for a set of FASTQ files including per cycle quality box plots, base proportions, base-level quality trends, relative k-mer diversity, length, and occurrence distribution of reads, number of reads above quality cutoffs and mean quality distribution. The results are written to a PDF file named <code>fastqReport.pdf</code>.</p>
<p>This is the pre-trimming fastq report. Another post-trimming fastq report step is not included in the default. It is recommended to run this step first to decide whether the trimming is needed.</p>
<p>Please note that initial targets files are being used here. In this case, it has been added to the first step, and later, we used the function <code>getColumn</code> to extract a named vector.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"targetsPE.txt"</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span>
    <span class="fu">updateColumn</span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"load_SPR"</span>, position <span class="op">=</span> <span class="st">"targetsWF"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">targets</span>
    <span class="va">fq_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"load_SPR"</span>, <span class="st">"targetsWF"</span>, column <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
    <span class="va">fqlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html" class="external-link">seeFastq</a></span><span class="op">(</span>fastq <span class="op">=</span> <span class="va">fq_files</span>, batchsize <span class="op">=</span> <span class="fl">10000</span>, klength <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span><span class="st">"./results/fastqReport_pre.pdf"</span>, height <span class="op">=</span> <span class="fl">18</span>, width <span class="op">=</span> <span class="fl">4</span> <span class="op">*</span>
        <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">fqlist</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html" class="external-link">seeFastqPlot</a></span><span class="op">(</span><span class="va">fqlist</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"fastq_report_pre"</span>, dependency <span class="op">=</span> <span class="st">"load_SPR"</span><span class="op">)</span></code></pre></div>
<img src="results/fastqReport.png"><div align="center">
Figure 1: FASTQ quality report for 18 samples
</div>
<p></p>
</div>
<div class="section level3">
<h3 id="read-preprocessing">Read preprocessing<a class="anchor" aria-label="anchor" href="#read-preprocessing"></a>
</h3>
<div class="section level4">
<h4 id="read-trimming-with-trimmomatic">Read trimming with Trimmomatic<a class="anchor" aria-label="anchor" href="#read-trimming-with-trimmomatic"></a>
</h4>
<p>Next, we need to populate the object created with the first step in the workflow. Here, an example of how to perform this task using parameters template files for trimming FASTQ files with <a href="http://www.usadellab.org/cms/?page=trimmomatic" class="external-link">Trimmomatic</a> software <span class="citation">(Bolger, Lohse, and Usadel 2014)</span>. For this step, the <code>SYSargsList</code> function has been used to build the command-line and append to <code>sal</code> object. For more details of all the features and utilities, please consult the <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html" class="external-link">main vignette</a>.</p>
<p>If GATK (default) is used for variant calling, any type of fastq trimming is strongly depreciated. GATK have internal function to handle low quality posistions.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"trimmomatic"</span>, targets <span class="op">=</span> <span class="st">"targetsPE.txt"</span>,
    wf_file <span class="op">=</span> <span class="st">"trimmomatic/trimmomatic-pe.cwl"</span>, input_file <span class="op">=</span> <span class="st">"trimmomatic/trimmomatic-pe.yml"</span>,
    dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>,
        FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>,
    dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fastq_report_pre"</span><span class="op">)</span>, run_step <span class="op">=</span> <span class="st">"optional"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="preprocessing-with-preprocessreads-function">Preprocessing with <em><code>preprocessReads</code></em> function<a class="anchor" aria-label="anchor" href="#preprocessing-with-preprocessreads-function"></a>
</h4>
<p>The function <em><code>preprocessReads</code></em> allows to apply predefined or custom read preprocessing functions to all FASTQ files referenced in a <em><code>SYSargsList</code></em> container, such as quality filtering or adaptor trimming routines. Internally, <em><code>preprocessReads</code></em> uses the <em><code>FastqStreamer</code></em> function from the <em><code>ShortRead</code></em> package to stream through large FASTQ files in a memory-efficient manner. The following example performs adaptor trimming with the <em><code>trimLRPatterns</code></em> function from the <em><code>Biostrings</code></em> package.</p>
<p>Here, we are appending this step at the <em><code>SYSargsList</code></em> object created previously. All the parameters are defined on the <em><code>preprocessReads/preprocessReads-pe.yml</code></em> file.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"preprocessing"</span>, targets <span class="op">=</span> <span class="st">"targetsPE.txt"</span>,
    dir <span class="op">=</span> <span class="cn">TRUE</span>, wf_file <span class="op">=</span> <span class="st">"preprocessReads/preprocessReads-pe.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"preprocessReads/preprocessReads-pe.yml"</span>, dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>,
    inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>,
        SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>, dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fastq_report_pre"</span><span class="op">)</span>,
    run_step <span class="op">=</span> <span class="st">"optional"</span><span class="op">)</span></code></pre></div>
<p>After the trimming step, the <code>outfiles</code> files can be used to generate the new targets files containing the paths to the trimmed FASTQ files. The new targets information can be used for the next workflow step instance, <em>e.g.</em> running the NGS alignments with the trimmed FASTQ files.</p>
<p>The following example shows how one can design a custom read <em>‘preprocessReads’</em> function using utilities provided by the <em><code>ShortRead</code></em> package, and then run it in batch mode with the <em>‘preprocessReads’</em> function. For here, it is possible to replace the function used on the <code>preprocessing</code> step and modify the <code>sal</code> object. Because it is a custom function, it is necessary to save the part in the R object, and internally the <code>preprocessReads.doc.R</code> is loading the function. If the R object is saved with a different name (here <code>"param/customFCT.RData"</code>), please replace that accordingly in the <code>preprocessReads.doc.R</code>.</p>
<p>Please, note that this step is not added to the workflow, here just for demonstration.</p>
<p>First, we defined the function in the workflow:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">filterFct</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">fq</span>, <span class="va">cutoff</span> <span class="op">=</span> <span class="fl">20</span>, <span class="va">Nexceptions</span> <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span>
        <span class="va">qcount</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/QualityScaledXStringSet-class.html" class="external-link">quality</a></span><span class="op">(</span><span class="va">fq</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="va">cutoff</span>,
            na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
        <span class="co"># Retains reads where Phred scores are &gt;= cutoff</span>
        <span class="co"># with N exceptions</span>
        <span class="va">fq</span><span class="op">[</span><span class="va">qcount</span> <span class="op">&lt;=</span> <span class="va">Nexceptions</span><span class="op">]</span>
    <span class="op">}</span>
    <span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span>, file <span class="op">=</span> <span class="st">"param/customFCT.RData"</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"custom_preprocessing_function"</span>, dependency <span class="op">=</span> <span class="st">"preprocessing"</span><span class="op">)</span></code></pre></div>
<p>After, we can edit the input parameter:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">yamlinput</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"preprocessing"</span><span class="op">)</span><span class="op">$</span><span class="va">Fct</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">yamlinput</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"preprocessing"</span>, <span class="st">"Fct"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"'filterFct(fq, cutoff=20, Nexceptions=0)'"</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">yamlinput</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"preprocessing"</span><span class="op">)</span><span class="op">$</span><span class="va">Fct</span>  <span class="co">## check the new function</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">cmdlist</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"preprocessing"</span>, targets <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>  <span class="co">## check if the command line was updated with success</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="fastq-quality-after-trimming">FASTQ quality after Trimming<a class="anchor" aria-label="anchor" href="#fastq-quality-after-trimming"></a>
</h3>
<p>This is the post-trimming fastq quality report. If the trimming step is included, it is recommended to add this step to compare trimming of fastq before and after.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">fq_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"preprocessing"</span>, <span class="st">"outfiles"</span>, column <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>  <span class="co">## get outfiles path</span>
    <span class="va">fqlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html" class="external-link">seeFastq</a></span><span class="op">(</span>fastq <span class="op">=</span> <span class="va">fq_files</span>, batchsize <span class="op">=</span> <span class="fl">10000</span>, klength <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span><span class="st">"./results/fastqReport_pos.pdf"</span>, height <span class="op">=</span> <span class="fl">18</span>, width <span class="op">=</span> <span class="fl">4</span> <span class="op">*</span>
        <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">fqlist</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html" class="external-link">seeFastqPlot</a></span><span class="op">(</span><span class="va">fqlist</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"fastq_report_pos"</span>, dependency <span class="op">=</span> <span class="st">"trimmomatic"</span>,
    run_step <span class="op">=</span> <span class="st">"optional"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="read-mapping-with-bwa-mem">Read mapping with <code>BWA-MEM</code><a class="anchor" aria-label="anchor" href="#read-mapping-with-bwa-mem"></a>
</h3>
<p>The NGS reads of this project are aligned against the reference genome sequence using the highly variant tolerant short read aligner <code>BWA-MEM</code> <span class="citation">(Heng Li 2013; H. Li and Durbin 2009)</span>. The parameter settings of the aligner are defined in the <code>param/cwl/gatk/bwa-pe.cwl</code>.</p>
<p>This test code uses untrimmed fastq files since the demo data is minimal and limited. However, it is best to test with <code>FASTQ quality report</code> function provided above to verify your real data first.</p>
<div class="section level4">
<h4 id="build-index-and-dictionary-files-for-bwa-and-gatk">Build index and dictionary files for BWA and GATK<a class="anchor" aria-label="anchor" href="#build-index-and-dictionary-files-for-bwa-and-gatk"></a>
</h4>
<p>Build the index and dictionary files for BWA and GATK to run.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"bwa_index"</span>, dir <span class="op">=</span> <span class="cn">FALSE</span>,
    targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"gatk/workflow_bwa-index.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>, dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, dependency <span class="op">=</span> <span class="st">"load_SPR"</span><span class="op">)</span></code></pre></div>
<p>Create reference <code>fasta</code> dictionary.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"fasta_index"</span>, dir <span class="op">=</span> <span class="cn">FALSE</span>,
    targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"gatk/workflow_fasta_dict.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>, dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, dependency <span class="op">=</span> <span class="st">"bwa_index"</span><span class="op">)</span></code></pre></div>
<p>Create dictionary index.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"faidx_index"</span>, dir <span class="op">=</span> <span class="cn">FALSE</span>,
    targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"gatk/workflow_fasta_faidx.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>, dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, dependency <span class="op">=</span> <span class="st">"fasta_index"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="mapping-reads-with-bwa">Mapping reads with BWA<a class="anchor" aria-label="anchor" href="#mapping-reads-with-bwa"></a>
</h4>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"bwa_alignment"</span>, targets <span class="op">=</span> <span class="st">"targetsPE.txt"</span>,
    wf_file <span class="op">=</span> <span class="st">"gatk/workflow_bwa-pe.cwl"</span>, input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>,
    dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>,
        FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>,
    dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"faidx_index"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="read-and-alignment-stats">Read and alignment stats<a class="anchor" aria-label="anchor" href="#read-and-alignment-stats"></a>
</h3>
<p>The following provides an overview of the number of reads in each sample and how many of them aligned to the reference.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">bampaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"bwa_alignment"</span>, <span class="st">"outfiles"</span>,
        column <span class="op">=</span> <span class="st">"samtools_sort_bam"</span><span class="op">)</span>
    <span class="va">fqpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"bwa_alignment"</span>, <span class="st">"targetsWF"</span>,
        column <span class="op">=</span> <span class="st">"FileName1"</span><span class="op">)</span>
    <span class="va">read_statsDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/alignStats.html" class="external-link">alignStats</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">bampaths</span>, fqpaths <span class="op">=</span> <span class="va">fqpaths</span>,
        pairEnd <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">read_statsDF</span>, <span class="st">"results/alignStats.xls"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>,
        quote <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"align_stats"</span>, dependency <span class="op">=</span> <span class="st">"bwa_alignment"</span>, run_step <span class="op">=</span> <span class="st">"optional"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-symbolic-links-for-viewing-bam-files-in-igv">Create symbolic links for viewing BAM files in IGV<a class="anchor" aria-label="anchor" href="#create-symbolic-links-for-viewing-bam-files-in-igv"></a>
</h3>
<p>The <code>symLink2bam</code> function creates symbolic links to view the BAM alignment files in a genome browser such as IGV. The corresponding URLs are written to a file with a path specified under <code>urlfile</code> in the <code>results</code> directory.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">bampaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"bwa_alignment"</span>, <span class="st">"outfiles"</span>,
        column <span class="op">=</span> <span class="st">"samtools_sort_bam"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/symLink2bam.html" class="external-link">symLink2bam</a></span><span class="op">(</span>sysargs <span class="op">=</span> <span class="va">bampaths</span>, htmldir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/.html/"</span>, <span class="st">"somedir/"</span><span class="op">)</span>,
        urlbase <span class="op">=</span> <span class="st">"http://cluster.hpcc.ucr.edu/~tgirke/"</span>, urlfile <span class="op">=</span> <span class="st">"./results/IGVurl.txt"</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"bam_urls"</span>, dependency <span class="op">=</span> <span class="st">"bwa_alignment"</span>, run_step <span class="op">=</span> <span class="st">"optional"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="variant-calling">Variant calling<a class="anchor" aria-label="anchor" href="#variant-calling"></a>
</h3>
<p>The following performs variant calling with <code>GATK</code> and <code>BCFtools</code> on a single machine by <code>runWF</code> function for each sample sequentially. If a cluster compute is available, running in parallel mode on a compute cluster can be performed by <code>runWF</code>, making available the resources and choose <code>run_session = "compute"</code>.</p>
<p>Not all users have a cluster system, so here to demonstrate an example of variant calling workflow, only single-machine commands are shown. For cluster jobs, please refer to our main <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html" class="external-link">vignette</a>.</p>
<p>In addition, the user would choose only one variant caller here rather than running several ones. However, the workflow manager allows keeping multiple options available for running the analysis.</p>
<div class="section level4">
<h4 id="variant-calling-with-gatk">Variant calling with <code>GATK</code><a class="anchor" aria-label="anchor" href="#variant-calling-with-gatk"></a>
</h4>
<p>The following steps are based on <code>GATK 4.1.1.0</code> <a href="https://software.broadinstitute.org/gatk/best-practices/" class="external-link">Best Practice</a>. There are 10 individual steps where the user can choose where to jump in and where to skip. All scripts are located at <code>param/cwl/gatk</code>. <code>BQSR</code> (Base Quality Score Recalibration) and <code>VQSR</code> (Variant Quality Score Recalibration) are very specific to a limited species like human, so this workflow does not support these steps.</p>
</div>
<div class="section level4">
<h4 id="step1-fastq-to-ubam">Step1: <code>fastq</code> to <code>ubam</code><a class="anchor" aria-label="anchor" href="#step1-fastq-to-ubam"></a>
</h4>
<p>Convert <code>fastq</code> files to <code>bam</code> files to prepare for the following step. It is very important to specific your sequencing platform, default is <code>illumina</code>. User need to change <code>param/cwl/gatk/gatk_fastq2ubam.cwl</code> if the platform is different. Platform information is needed for the variant caller in later steps to correct calling parameters.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"fastq2ubam"</span>, targets <span class="op">=</span> <span class="st">"targetsPE.txt"</span>,
    wf_file <span class="op">=</span> <span class="st">"gatk/workflow_gatk_fastq2ubam.cwl"</span>, input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>,
    dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>,
        FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>,
    dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"faidx_index"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step2-merge-bam-and-ubam">Step2: Merge <code>bam</code> and <code>ubam</code><a class="anchor" aria-label="anchor" href="#step2-merge-bam-and-ubam"></a>
</h4>
<p>This step merges a <code>bam</code> and <code>ubam</code> and creates a third <code>bam</code> file that contains alignment information and remaining information that was removed by the aligner like <code>BWA</code>. The removed information is essential for variant statistics calculation. Previous steps are recommended, but variant calling can still be performed without these steps.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"merge_bam"</span>, targets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bwa_alignment"</span>,
    <span class="st">"fastq2ubam"</span><span class="op">)</span>, wf_file <span class="op">=</span> <span class="st">"gatk/workflow_gatk_mergebams.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>, dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>bwa_men_sam <span class="op">=</span> <span class="st">"_bwasam_"</span>,
        ubam <span class="op">=</span> <span class="st">"_ubam_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>, rm_targets_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"preprocessReads_1"</span>,
        <span class="st">"preprocessReads_2"</span><span class="op">)</span>, dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bwa_alignment"</span>,
        <span class="st">"fastq2ubam"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step3-sort-bam-files-by-genomic-coordinates">Step3: Sort <code>bam</code> files by genomic coordinates<a class="anchor" aria-label="anchor" href="#step3-sort-bam-files-by-genomic-coordinates"></a>
</h4>
<p>Sort <code>bam</code> files by genomic coordinates.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"sort"</span>, targets <span class="op">=</span> <span class="st">"merge_bam"</span>,
    wf_file <span class="op">=</span> <span class="st">"gatk/workflow_gatk_sort.cwl"</span>, input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>,
    dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>merge_bam <span class="op">=</span> <span class="st">"_mergebam_"</span>,
        SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>, rm_targets_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bwa_men_sam"</span>,
        <span class="st">"ubam"</span>, <span class="st">"SampleName_fastq2ubam"</span>, <span class="st">"Factor_fastq2ubam"</span>,
        <span class="st">"SampleLong_fastq2ubam"</span>, <span class="st">"Experiment_fastq2ubam"</span>, <span class="st">"Date_fastq2ubam"</span><span class="op">)</span>,
    dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"merge_bam"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step4-mark-duplicates">Step4: Mark duplicates<a class="anchor" aria-label="anchor" href="#step4-mark-duplicates"></a>
</h4>
<p>Mark PCR artifacts in sequencing. A <code>duplicate_metrics</code> file will also be produced by this step, but will not be used for the next step. This file is just for the user to check duplicates status summary.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"mark_dup"</span>, targets <span class="op">=</span> <span class="st">"sort"</span>,
    wf_file <span class="op">=</span> <span class="st">"gatk/workflow_gatk_markduplicates.cwl"</span>, input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>,
    dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>sort_bam <span class="op">=</span> <span class="st">"_sort_"</span>,
        SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>, rm_targets_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"merge_bam"</span><span class="op">)</span>,
    dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sort"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step5-fixing-tags">Step5: Fixing tags<a class="anchor" aria-label="anchor" href="#step5-fixing-tags"></a>
</h4>
<p>Takes the <code>bam</code> from the last step and calculates the NM, MD, and UQ tags. These tags are important for variant calling and filtering. This step is recommended but can be skipped.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"fix_tag"</span>, targets <span class="op">=</span> <span class="st">"mark_dup"</span>,
    wf_file <span class="op">=</span> <span class="st">"gatk/workflow_gatk_fixtag.cwl"</span>, input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>,
    dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>mark_bam <span class="op">=</span> <span class="st">"_mark_"</span>,
        SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>, rm_targets_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sort_bam"</span><span class="op">)</span>,
    dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mark_dup"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Up till this step, sample preprocess is done. All analysis ready <code>BAM</code> files and their index <code>.bai</code> files are created. Individual and cohort calling by <code>HaplotypeCaller</code> is performed from the next step.</p>
</div>
<div class="section level4">
<h4 id="step6-haplotypecaller-gvcf">Step6: HaplotypeCaller <code>gvcf</code><a class="anchor" aria-label="anchor" href="#step6-haplotypecaller-gvcf"></a>
</h4>
<p>The <code>HaplotypeCaller</code> is running a <strong>gvcf</strong> mode in this step. G stands for ‘genomic.’ The file not only contains variant sites information but also non-variant sites information; thus, at the following step, the cohort caller can use this information to validate the true variants.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"hap_caller"</span>, targets <span class="op">=</span> <span class="st">"fix_tag"</span>,
    wf_file <span class="op">=</span> <span class="st">"gatk/workflow_gatk_haplotypecaller.cwl"</span>, input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>,
    dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>fixtag_bam <span class="op">=</span> <span class="st">"_fixed_"</span>,
        SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>, rm_targets_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mark_bam"</span><span class="op">)</span>,
    dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fix_tag"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step7-import-all-gvcfs">Step7: Import all <code>gvcfs</code><a class="anchor" aria-label="anchor" href="#step7-import-all-gvcfs"></a>
</h4>
<p>It is recommended to import all <strong>gvcfs</strong> to a <a href="https://github.com/Intel-HLS/GenomicsDB/wiki" class="external-link">TileDB</a> database for fast cohort variant calling at the following step. Note: if you are working with non-diploid data, use <code>CombineGVCFs</code> function from <code>GATK</code> and change the <code>gvcf_db_folder</code> parameter in <code>param/cwl/gatk/gatk.yaml</code> to be your combined <strong>gvcf</strong> file path.</p>
<p><strong>Important</strong>: Make sure all samples’ <code>*.g.vcf.gz</code> files are in the results folder, also the <code>tbi index</code> files also should be there.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"import"</span>, targets <span class="op">=</span> <span class="cn">NULL</span>,
    dir <span class="op">=</span> <span class="cn">FALSE</span>, wf_file <span class="op">=</span> <span class="st">"gatk/workflow_gatk_genomicsDBImport.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>, dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hap_caller"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step8-cohort-calling-of-gvcf">Step8: Cohort calling of <code>gvcf</code><a class="anchor" aria-label="anchor" href="#step8-cohort-calling-of-gvcf"></a>
</h4>
<p>Assess variants by information from all <code>gvcfs</code>. A collective <code>vcf</code> called <code>samples.vcf.gz</code> is created by default naming.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"call_variants"</span>, targets <span class="op">=</span> <span class="cn">NULL</span>,
    dir <span class="op">=</span> <span class="cn">FALSE</span>, wf_file <span class="op">=</span> <span class="st">"gatk/workflow_gatk_genotypeGVCFs.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>, dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"import"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step9-cohort-hard-filter-variants">Step9: Cohort hard filter variants<a class="anchor" aria-label="anchor" href="#step9-cohort-hard-filter-variants"></a>
</h4>
<p>Variant Quality Score Recalibration (VQSR) is not included in this workflow. Variants are hard filtered together. See this <a href="https://gatkforums.broadinstitute.org/gatk/discussion/2806/howto-apply-hard-filters-to-a-call-set" class="external-link">Post</a> for parameters for hard filtering. Change these settings in <code>param/cwl/gak/gatk_variantFiltration.sh</code> if needed. VQSR requires a large quantity of samples to be training data before you can do filtering. Read this <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035531612-Variant-Quality-Score-Recalibration-VQSR-" class="external-link">post</a> for more information.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"filter"</span>, targets <span class="op">=</span> <span class="cn">NULL</span>,
    dir <span class="op">=</span> <span class="cn">FALSE</span>, wf_file <span class="op">=</span> <span class="st">"gatk/workflow_gatk_variantFiltration.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>, dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"call_variants"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step10-extract-variant">Step10: Extract variant<a class="anchor" aria-label="anchor" href="#step10-extract-variant"></a>
</h4>
<p>After cohort calling, filtering, all variants for all samples are stored in one big file. Extract variants for each sample and save them separately (only variants that have passed the filters are stored).</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"create_vcf"</span>, targets <span class="op">=</span> <span class="st">"hap_caller"</span>,
    wf_file <span class="op">=</span> <span class="st">"gatk/workflow_gatk_select_variant.cwl"</span>, input_file <span class="op">=</span> <span class="st">"gatk/gatk.yaml"</span>,
    dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>,
    dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hap_caller"</span>, <span class="st">"filter"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="variant-calling-with-bcftools">Variant calling with <code>BCFtools</code><a class="anchor" aria-label="anchor" href="#variant-calling-with-bcftools"></a>
</h3>
<p>Alternative option with <code>BCFtool</code>:</p>
<p>The following runs the variant calling with <code>BCFtools</code>. This tool takes <code>BWA</code> aligned <code>BAM</code> files, sort, mark duplicates by <code>samtools</code> and finally call variants by <code>BCFtools</code>. <strong>For legacy reasons we keep this option.</strong></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"create_vcf_BCFtool"</span>,
    targets <span class="op">=</span> <span class="st">"bwa_alignment"</span>, dir <span class="op">=</span> <span class="cn">TRUE</span>, wf_file <span class="op">=</span> <span class="st">"workflow-bcftools/workflow_bcftools.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"workflow-bcftools/bcftools.yml"</span>, dir_path <span class="op">=</span> <span class="st">"param/cwl"</span>,
    inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>bwa_men_sam <span class="op">=</span> <span class="st">"_bwasam_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>,
    rm_targets_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"preprocessReads_1"</span>, <span class="st">"preprocessReads_2"</span><span class="op">)</span>,
    dependency <span class="op">=</span> <span class="st">"bwa_alignment"</span>, run_step <span class="op">=</span> <span class="st">"optional"</span><span class="op">)</span></code></pre></div>
<p>Variant calling ends here. Downstream analysis starts from the next section.</p>
</div>
<div class="section level3">
<h3 id="inspect-vcf-file">Inspect VCF file<a class="anchor" aria-label="anchor" href="#inspect-vcf-file"></a>
</h3>
<p>Scripts of downstream analysis are stored in <code>param/cwl/varseq_downstream</code>.</p>
<p><strong>optional</strong>: This step is not included in the default workflow. After successfully execute the entire workflow, users may load individual vcf files to R for other analysis like below.</p>
<p>VCF files can be imported into R with the <code>readVcf</code> function. Both <code>VCF</code> and <code>VRanges</code> objects provide convenient data structure for working with variant data (<em>e.g.</em> SNP quality filtering).</p>
<p>This step is not included in the default workflow steps, but can be useful to inspect individual sample’s raw variants.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VariantAnnotation</span><span class="op">)</span>
<span class="va">vcf_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"create_vcf"</span><span class="op">)</span>
<span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span><span class="va">vcf_raw</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"A. thaliana"</span><span class="op">)</span>
<span class="va">vcf</span>
<span class="va">vr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="st">"VRanges"</span><span class="op">)</span>
<span class="va">vr</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-variants">Filter variants<a class="anchor" aria-label="anchor" href="#filter-variants"></a>
</h3>
<p>The function <code>filterVars</code> filters VCF files based on user definable quality parameters. It sequentially imports each VCF file into R, applies the filtering on an internally generated <code>VRanges</code> object and then writes the results to a new subsetted VCF file. The filter parameters are passed on to the corresponding argument as a character string. The function applies this filter to the internally generated <code>VRanges</code> object using the standard subsetting syntax for two dimensional objects such as: <code>vr[filter, ]</code>.</p>
<div class="section level4">
<h4 id="filter-variants-called-by-gatk">Filter variants called by <code>GATK</code><a class="anchor" aria-label="anchor" href="#filter-variants-called-by-gatk"></a>
</h4>
<p>The below example filters for variants that are supported by <code>&gt;=x</code> reads and &gt;=80% of them support the called variants. In addition, all variants need to pass <code>&gt;=x</code> of the soft filters recorded in the VCF files generated by GATK. Since the toy data used for this workflow is very small, the chosen settings are unreasonabley relaxed. A more reasonable filter setting is given in the line below (here commented out).</p>
<p>There is already some cohort filtering in GATK step 10. Some additional hard filtering is provided here. This step is included here, but in a real analysis, you may skip this step.</p>
<p>For real samples, use following filters: <code>filter &lt;- "totalDepth(vr) &gt;= 20 &amp; (altDepth(vr) / totalDepth(vr) &gt;= 0.8)"</code></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">vcf_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"create_vcf"</span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VariantAnnotation</span><span class="op">)</span>
    <span class="va">filter</span> <span class="op">&lt;-</span> <span class="st">"totalDepth(vr) &gt;= 2 &amp; (altDepth(vr) / totalDepth(vr) &gt;= 0.8)"</span>
    <span class="va">vcf_filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/filterVars.html" class="external-link">filterVars</a></span><span class="op">(</span><span class="va">vcf_raw</span>, <span class="va">filter</span>,
        organism <span class="op">=</span> <span class="st">"A. thaliana"</span>, out_dir <span class="op">=</span> <span class="st">"results/vcf_filter"</span><span class="op">)</span><span class="op">)</span>
    <span class="co"># dump the filtered path variable to running</span>
    <span class="co"># enviornment so other sysArg steps can get its values</span>
    <span class="fu">updateColumn</span><span class="op">(</span><span class="va">sal</span>, <span class="st">"create_vcf"</span>, <span class="st">"outfiles"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>vcf_filter <span class="op">=</span> <span class="va">vcf_filter</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"filter_vcf"</span>, dependency <span class="op">=</span> <span class="st">"create_vcf"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="filter-variants-called-by-bcftools">Filter variants called by <code>BCFtools</code><a class="anchor" aria-label="anchor" href="#filter-variants-called-by-bcftools"></a>
</h4>
<p>The following shows how to filter the VCF files generated by <code>BCFtools</code> using similar parameter settings as in the previous filtering of the GATK results.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">vcf_raw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"create_vcf_BCFtool"</span>, position <span class="op">=</span> <span class="st">"outfiles"</span>,
        column <span class="op">=</span> <span class="st">"bcftools_call"</span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VariantAnnotation</span><span class="op">)</span>
    <span class="va">filter</span> <span class="op">&lt;-</span> <span class="st">"rowSums(vr) &gt;= 2 &amp; (rowSums(vr[,3:4])/rowSums(vr[,1:4]) &gt;= 0.8)"</span>
    <span class="va">vcf_filter_bcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/filterVars.html" class="external-link">filterVars</a></span><span class="op">(</span><span class="va">vcf_raw</span>, <span class="va">filter</span>,
        organism <span class="op">=</span> <span class="st">"A. thaliana"</span>, out_dir <span class="op">=</span> <span class="st">"results/vcf_filter_BCFtools"</span>,
        varcaller <span class="op">=</span> <span class="st">"bcftools"</span><span class="op">)</span><span class="op">)</span>

    <span class="fu">updateColumn</span><span class="op">(</span><span class="va">sal</span>, <span class="st">"create_vcf"</span>, <span class="st">"outfiles"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>vcf_filter_bcf <span class="op">=</span> <span class="va">vcf_filter_bcf</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"filter_vcf_BCFtools"</span>, dependency <span class="op">=</span> <span class="st">"create_vcf_BCFtool"</span>,
    run_step <span class="op">=</span> <span class="st">"optional"</span><span class="op">)</span></code></pre></div>
<p>Check filtering outcome for one sample</p>
<p>This mini step can be used to compare <code>vcfs</code> files before and after filtering. This can be used once the workflow has been run, and make sure “filter_vcf” is done, since it is an optional step.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">copyEnvir</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"vcf_raw"</span>, <span class="fu"><a href="https://rdrr.io/r/base/environment.html" class="external-link">globalenv</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">copyEnvir</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"vcf_filter"</span>, <span class="fu"><a href="https://rdrr.io/r/base/environment.html" class="external-link">globalenv</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span><span class="va">vcf_raw</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, genome <span class="op">=</span> <span class="st">"Ath"</span><span class="op">)</span>, <span class="st">"VRanges"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span><span class="va">vcf_filter</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, genome <span class="op">=</span> <span class="st">"Ath"</span><span class="op">)</span>, <span class="st">"VRanges"</span><span class="op">)</span><span class="op">[</span>,
    <span class="fl">1</span><span class="op">]</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="annotate-filtered-variants">Annotate filtered variants<a class="anchor" aria-label="anchor" href="#annotate-filtered-variants"></a>
</h3>
<p>The function <code>variantReport</code> generates a variant report using utilities provided by the <code>VariantAnnotation</code> package. The report for each sample is written to a tabular file containing genomic context annotations (<em>e.g.</em> coding or non-coding SNPs, amino acid changes, IDs of affected genes, etc.) along with confidence statistics for each variant. The CWL file <code>param/cwl/varseq_downstream/annotate.cwl</code> defines the paths to the input and output files which are stored in a <code>SYSargs2</code> instance.</p>
<div class="section level4">
<h4 id="basics-of-annotating-variants">Basics of annotating variants<a class="anchor" aria-label="anchor" href="#basics-of-annotating-variants"></a>
</h4>
<p>This step can be run after running the default workflow, not included in the default.</p>
<p>Variants overlapping with common annotation features can be identified with <code>locateVariants</code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">"GenomicFeatures"</a></span><span class="op">)</span>
<span class="co"># comment the next line if optional step 'filter_vcf' is</span>
<span class="co"># included</span>
<span class="va">vcf_filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"create_vcf"</span><span class="op">)</span>
<span class="co"># uncomment the next line if optional step 'filter_vcf' is</span>
<span class="co"># included copyEnvir(sal, 'vcf_filter', globalenv())</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="st">"./data/tair10.sqlite"</span><span class="op">)</span>
<span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span><span class="va">vcf_filter</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"A. thaliana"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/locateVariants-methods.html" class="external-link">locateVariants</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">txdb</span>, <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VariantType-class.html" class="external-link">CodingVariants</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Synonymous/non-synonymous variants of coding sequences are computed by the <code>predictCoding</code> function for variants overlapping with coding regions.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/FaFile-class.html" class="external-link">FaFile</a></span><span class="op">(</span><span class="st">"data/tair10.fasta"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/predictCoding-methods.html" class="external-link">predictCoding</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="va">txdb</span>, seqSource <span class="op">=</span> <span class="va">fa</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="annotate-filtered-variants-gatk-or-bcftools">Annotate filtered variants <code>GATK</code> or <code>BCFtools</code><a class="anchor" aria-label="anchor" href="#annotate-filtered-variants-gatk-or-bcftools"></a>
</h4>
<p><strong>required</strong></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="co"># get the filtered vcf path from R running environment</span>
    <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">copyEnvir</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"vcf_filter"</span>, <span class="fu"><a href="https://rdrr.io/r/base/environment.html" class="external-link">globalenv</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">"GenomicFeatures"</a></span><span class="op">)</span>
    <span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="st">"./data/tair10.sqlite"</span><span class="op">)</span>
    <span class="va">fa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/FaFile-class.html" class="external-link">FaFile</a></span><span class="op">(</span><span class="st">"data/tair10.fasta"</span><span class="op">)</span>
    <span class="va">vcf_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/variantReport.html" class="external-link">variantReport</a></span><span class="op">(</span><span class="va">vcf_filter</span>,
        txdb <span class="op">=</span> <span class="va">txdb</span>, fa <span class="op">=</span> <span class="va">fa</span>, organism <span class="op">=</span> <span class="st">"A. thaliana"</span>, out_dir <span class="op">=</span> <span class="st">"results/vcf_anno"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"annotate_vcf"</span>, dependency <span class="op">=</span> <span class="st">"filter_vcf"</span><span class="op">)</span></code></pre></div>
<p>View annotation result for single sample</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">copyEnvir</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"vcf_anno"</span>, <span class="fu"><a href="https://rdrr.io/r/base/environment.html" class="external-link">globalenv</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">vcf_anno</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">38</span><span class="op">:</span><span class="fl">40</span>, <span class="op">]</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="combine-annotation-results-among-samples">Combine annotation results among samples<a class="anchor" aria-label="anchor" href="#combine-annotation-results-among-samples"></a>
</h3>
<p>To simplify comparisons among samples, the <code>combineVarReports</code> function combines all variant annotation reports referenced in a <code>SYSargs2</code> instance (here <code>args</code>). At the same time the function allows to consider only certain feature types of interest. For instance, the below setting <code>filtercol=c(Consequence="nonsynonymous")</code> will include only nonsysynonymous variances listed in the <code>Consequence</code> column of the annotation reports. To omit filtering, one can use the setting <code>filtercol="All"</code>.</p>
<div class="section level4">
<h4 id="combine-results">Combine results<a class="anchor" aria-label="anchor" href="#combine-results"></a>
</h4>
<p><strong>required</strong></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">combineDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/variantReport.html" class="external-link">combineVarReports</a></span><span class="op">(</span><span class="va">vcf_anno</span>, filtercol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>Consequence <span class="op">=</span> <span class="st">"nonsynonymous"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">combineDF</span>, <span class="st">"./results/combineDF_nonsyn.tsv"</span>,
        quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"combine_var"</span>, dependency <span class="op">=</span> <span class="st">"annotate_vcf"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="summary-statistics-of-variants">Summary statistics of variants<a class="anchor" aria-label="anchor" href="#summary-statistics-of-variants"></a>
</h3>
<p>The <code>varSummary</code> function counts the number of variants for each feature type included in the annotation reports.</p>
</div>
<div class="section level3">
<h3 id="summary-of-variants">Summary of variants<a class="anchor" aria-label="anchor" href="#summary-of-variants"></a>
</h3>
<p><strong>required</strong></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/variantReport.html" class="external-link">varSummary</a></span><span class="op">(</span><span class="va">vcf_anno</span><span class="op">)</span>, <span class="st">"./results/variantStats.tsv"</span>,
        quote <span class="op">=</span> <span class="cn">FALSE</span>, col.names <span class="op">=</span> <span class="cn">NA</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"summary_var"</span>, dependency <span class="op">=</span> <span class="st">"combine_var"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="venn-diagram-of-variants">Venn diagram of variants<a class="anchor" aria-label="anchor" href="#venn-diagram-of-variants"></a>
</h3>
<p><strong>Optional</strong> but included in the default</p>
<p>The venn diagram utilities defined by the <code>systemPipeR</code> package can be used to identify common and unique variants reported for different samples and/or variant callers. The below generates a 3-way venn diagram comparing 3 samples for each of the two variant callers.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="co">## make a list of first three samples</span>
    <span class="va">varlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">vcf_anno</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">vcf_anno</span><span class="op">[</span><span class="va">x</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">VARID</span><span class="op">)</span><span class="op">)</span>
    <span class="va">vennset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/overLapper.html" class="external-link">overLapper</a></span><span class="op">(</span><span class="va">varlist</span>, type <span class="op">=</span> <span class="st">"vennsets"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span><span class="st">"./results/vennplot_var.pdf"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/vennPlot.html" class="external-link">vennPlot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">vennset</span><span class="op">)</span>, mymain <span class="op">=</span> <span class="st">"Venn Plot of First 3 Samples"</span>,
        mysub <span class="op">=</span> <span class="st">""</span>, colmode <span class="op">=</span> <span class="fl">2</span>, ccol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"venn_diagram"</span>, dependency <span class="op">=</span> <span class="st">"annotate_vcf"</span><span class="op">)</span></code></pre></div>
<img src="results/vennplot_var.png"><div align="center">
Figure 2: Venn Diagram for 3 samples from GATK and BCFtools
</div>
<p></p>
</div>
<div class="section level3">
<h3 id="plot-variants-programmatically">Plot variants programmatically<a class="anchor" aria-label="anchor" href="#plot-variants-programmatically"></a>
</h3>
<p><strong>Optional</strong> but included in default</p>
<p>The following plots a selected variant with <code>ggbio</code>.</p>
<p>In this example, the input <code>BAM</code> file is from the <code>GATK</code> step 5, analysis ready bam. You can use other aligned <code>BAMs</code> as well, but make sure they are indexed. The <code>VCF</code> file is taken from <code>Inspect VCF file</code> section or you can load your own vcf.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="co"># get the filtered vcf path from R running environment</span>
    <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">copyEnvir</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"vcf_filter"</span>, <span class="fu"><a href="https://rdrr.io/r/base/environment.html" class="external-link">globalenv</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://lawremi.github.io/ggbio/" class="external-link">ggbio</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">VariantAnnotation</span><span class="op">)</span>
    <span class="va">mychr</span> <span class="op">&lt;-</span> <span class="st">"ChrM"</span>
    <span class="va">mystart</span> <span class="op">&lt;-</span> <span class="fl">19000</span>
    <span class="va">myend</span> <span class="op">&lt;-</span> <span class="fl">21000</span>
    <span class="va">bams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"fix_tag"</span><span class="op">)</span>
    <span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span><span class="va">vcf_filter</span><span class="op">[</span><span class="st">"M6B"</span><span class="op">]</span>, <span class="st">"A. thaliana"</span><span class="op">)</span><span class="op">)</span>
    <span class="va">ga</span> <span class="op">&lt;-</span> <span class="fu">readGAlignments</span><span class="op">(</span><span class="va">bams</span><span class="op">[</span><span class="st">"M6B"</span><span class="op">]</span>, use.names <span class="op">=</span> <span class="cn">TRUE</span>, param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html" class="external-link">ScanBamParam</a></span><span class="op">(</span>which <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="va">mychr</span>,
        <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="va">mystart</span>, <span class="va">myend</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
    <span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggbio/man/autoplot-method.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">ga</span>, geom <span class="op">=</span> <span class="st">"rect"</span><span class="op">)</span>
    <span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggbio/man/autoplot-method.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">ga</span>, geom <span class="op">=</span> <span class="st">"line"</span>, stat <span class="op">=</span> <span class="st">"coverage"</span><span class="op">)</span>
    <span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggbio/man/autoplot-method.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span> <span class="op">==</span> <span class="va">mychr</span><span class="op">]</span>, type <span class="op">=</span> <span class="st">"fixed"</span><span class="op">)</span> <span class="op">+</span>
        <span class="fu"><a href="https://rdrr.io/pkg/ggbio/man/tracks.html" class="external-link">xlim</a></span><span class="op">(</span><span class="va">mystart</span>, <span class="va">myend</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span>,
        axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
    <span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggbio/man/autoplot-method.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="st">"./data/tair10.sqlite"</span><span class="op">)</span>, which <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="va">mychr</span>,
        <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="va">mystart</span>, <span class="va">myend</span><span class="op">)</span><span class="op">)</span>, names.expr <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span>
    <span class="va">p1_4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggbio/man/tracks.html" class="external-link">tracks</a></span><span class="op">(</span>Reads <span class="op">=</span> <span class="va">p1</span>, Coverage <span class="op">=</span> <span class="va">p2</span>, Variant <span class="op">=</span> <span class="va">p3</span>, Transcripts <span class="op">=</span> <span class="va">p4</span>,
        heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.2</span>, <span class="fl">0.1</span>, <span class="fl">0.35</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span>
    <span class="fu">ggbio</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggbio/man/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="va">p1_4</span>, filename <span class="op">=</span> <span class="st">"./results/plot_variant.png"</span>,
        units <span class="op">=</span> <span class="st">"in"</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"plot_variant"</span>, dependency <span class="op">=</span> <span class="st">"filter_vcf"</span><span class="op">)</span></code></pre></div>
<img src="results/plot_variant.png"><div align="center">
Figure 3: Plot variants with programmatically.
</div>
<p></p>
</div>
<div class="section level3">
<h3 id="version-information">Version Information<a class="anchor" aria-label="anchor" href="#version-information"></a>
</h3>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R Under development (unstable) (2021-10-25 r81105)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.4 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span>
<span class="co">## LAPACK: /home/dcassol/src/R-devel/lib/libRlapack.so</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats4    stats     graphics  grDevices utils    </span>
<span class="co">## [6] datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] systemPipeR_2.1.30          ShortRead_1.53.1           </span>
<span class="co">##  [3] GenomicAlignments_1.31.2    SummarizedExperiment_1.25.3</span>
<span class="co">##  [5] Biobase_2.55.2              MatrixGenerics_1.7.0       </span>
<span class="co">##  [7] matrixStats_0.62.0          BiocParallel_1.29.21       </span>
<span class="co">##  [9] Rsamtools_2.11.0            Biostrings_2.63.3          </span>
<span class="co">## [11] XVector_0.35.0              GenomicRanges_1.47.6       </span>
<span class="co">## [13] GenomeInfoDb_1.31.10        IRanges_2.29.1             </span>
<span class="co">## [15] S4Vectors_0.33.17           BiocGenerics_0.41.2        </span>
<span class="co">## [17] BiocStyle_2.23.1           </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##  [1] sass_0.4.1             jsonlite_1.8.0        </span>
<span class="co">##  [3] bslib_0.3.1            assertthat_0.2.1      </span>
<span class="co">##  [5] BiocManager_1.30.17    latticeExtra_0.6-29   </span>
<span class="co">##  [7] GenomeInfoDbData_1.2.8 yaml_2.3.5            </span>
<span class="co">##  [9] pillar_1.7.0           lattice_0.20-45       </span>
<span class="co">## [11] glue_1.6.2             digest_0.6.29         </span>
<span class="co">## [13] RColorBrewer_1.1-3     colorspace_2.0-3      </span>
<span class="co">## [15] htmltools_0.5.2        Matrix_1.4-1          </span>
<span class="co">## [17] pkgconfig_2.0.3        bookdown_0.26         </span>
<span class="co">## [19] zlibbioc_1.41.0        purrr_0.3.4           </span>
<span class="co">## [21] scales_1.2.0           jpeg_0.1-9            </span>
<span class="co">## [23] tibble_3.1.6           generics_0.1.2        </span>
<span class="co">## [25] ggplot2_3.3.5          ellipsis_0.3.2        </span>
<span class="co">## [27] cachem_1.0.6           cli_3.3.0             </span>
<span class="co">## [29] magrittr_2.0.3         crayon_1.5.1          </span>
<span class="co">## [31] memoise_2.0.1          evaluate_0.15         </span>
<span class="co">## [33] fs_1.5.2               fansi_1.0.3           </span>
<span class="co">## [35] hwriter_1.3.2.1        textshaping_0.3.6     </span>
<span class="co">## [37] tools_4.2.0            formatR_1.12          </span>
<span class="co">## [39] lifecycle_1.0.1        stringr_1.4.0         </span>
<span class="co">## [41] munsell_0.5.0          DelayedArray_0.21.2   </span>
<span class="co">## [43] compiler_4.2.0         pkgdown_2.0.3         </span>
<span class="co">## [45] jquerylib_0.1.4        systemfonts_1.0.4     </span>
<span class="co">## [47] rlang_1.0.2            grid_4.2.0            </span>
<span class="co">## [49] RCurl_1.98-1.6         htmlwidgets_1.5.4     </span>
<span class="co">## [51] bitops_1.0-7           rmarkdown_2.14        </span>
<span class="co">## [53] codetools_0.2-18       gtable_0.3.0          </span>
<span class="co">## [55] DBI_1.1.2              R6_2.5.1              </span>
<span class="co">## [57] knitr_1.38             dplyr_1.0.8           </span>
<span class="co">## [59] fastmap_1.1.0          utf8_1.2.2            </span>
<span class="co">## [61] rprojroot_2.0.3        ragg_1.2.2            </span>
<span class="co">## [63] desc_1.4.1             stringi_1.7.6         </span>
<span class="co">## [65] parallel_4.2.0         vctrs_0.4.1           </span>
<span class="co">## [67] png_0.1-7              tidyselect_1.1.2      </span>
<span class="co">## [69] xfun_0.30</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="running-workflow">Running workflow<a class="anchor" aria-label="anchor" href="#running-workflow"></a>
</h2>
<div class="section level3">
<h3 id="interactive-job-submissions-in-a-single-machine">Interactive job submissions in a single machine<a class="anchor" aria-label="anchor" href="#interactive-job-submissions-in-a-single-machine"></a>
</h3>
<p>For running the workflow, <code>runWF</code> function will execute all the steps store in the workflow container. The execution will be on a single machine without submitting to a queuing system of a computer cluster.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html" class="external-link">runWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="parallelization-on-clusters">Parallelization on clusters<a class="anchor" aria-label="anchor" href="#parallelization-on-clusters"></a>
</h3>
<p>Alternatively, the computation can be greatly accelerated by processing many files in parallel using several compute nodes of a cluster, where a scheduling/queuing system is used for load balancing.</p>
<p>The <code>resources</code> list object provides the number of independent parallel cluster processes defined under the <code>Njobs</code> element in the list. The following example will run 18 processes in parallel using each 4 CPU cores. If the resources available on a cluster allow running all 18 processes at the same time, then the shown sample submission will utilize in a total of 72 CPU cores.</p>
<p>Note, <code>runWF</code> can be used with most queueing systems as it is based on utilities from the <code>batchtools</code> package, which supports the use of template files (<em><code>*.tmpl</code></em>) for defining the run parameters of different schedulers. To run the following code, one needs to have both a <code>conffile</code> (see <em><code>.batchtools.conf.R</code></em> samples <a href="https://mllg.github.io/batchtools/" class="external-link">here</a>) and a <code>template</code> file (see <em><code>*.tmpl</code></em> samples <a href="https://github.com/mllg/batchtools/tree/master/inst/templates" class="external-link">here</a>) for the queueing available on a system. The following example uses the sample <code>conffile</code> and <code>template</code> files for the Slurm scheduler provided by this package.</p>
<p>The resources can be appended when the step is generated, or it is possible to add these resources later, as the following example using the <code>addResources</code> function:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>conffile<span class="op">=</span><span class="st">".batchtools.conf.R"</span>,
                  template<span class="op">=</span><span class="st">"batchtools.slurm.tmpl"</span>, 
                  Njobs<span class="op">=</span><span class="fl">18</span>, 
                  walltime<span class="op">=</span><span class="fl">120</span>, <span class="co">## minutes</span>
                  ntasks<span class="op">=</span><span class="fl">1</span>,
                  ncpus<span class="op">=</span><span class="fl">4</span>, 
                  memory<span class="op">=</span><span class="fl">1024</span>, <span class="co">## Mb</span>
                  partition <span class="op">=</span> <span class="st">"short"</span>
                  <span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">addResources</a></span><span class="op">(</span><span class="va">sal</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hisat2_mapping"</span><span class="op">)</span>, resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html" class="external-link">runWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-workflow">Visualize workflow<a class="anchor" aria-label="anchor" href="#visualize-workflow"></a>
</h3>
<p><em><code>systemPipeR</code></em> workflows instances can be visualized with the <code>plotWF</code> function.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/plotWF.html" class="external-link">plotWF</a></span><span class="op">(</span><span class="va">sal</span>, rstudio <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="checking-workflow-status">Checking workflow status<a class="anchor" aria-label="anchor" href="#checking-workflow-status"></a>
</h3>
<p>To check the summary of the workflow, we can use:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">statusWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="accessing-logs-report">Accessing logs report<a class="anchor" aria-label="anchor" href="#accessing-logs-report"></a>
</h3>
<p><em><code>systemPipeR</code></em> compiles all the workflow execution logs in one central location, making it easier to check any standard output (<code>stdout</code>) or standard error (<code>stderr</code>) for any command-line tools used on the workflow or the R code stdout.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderLogs.html" class="external-link">renderLogs</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="funding">Funding<a class="anchor" aria-label="anchor" href="#funding"></a>
</h2>
<p>This project was supported by funds from the National Institutes of Health (NIH) and the National Science Foundation (NSF).</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bolger2014-yr" class="csl-entry">
Bolger, Anthony M, Marc Lohse, and Bjoern Usadel. 2014. <span>“Trimmomatic: A Flexible Trimmer for Illumina Sequence Data.”</span> <em>Bioinformatics</em> 30 (15): 2114–20.
</div>
<div id="ref-H_Backman2016-bt" class="csl-entry">
H Backman, Tyler W, and Thomas Girke. 2016. <span>“<span class="nocase">systemPipeR: NGS workflow and report generation environment</span>.”</span> <em>BMC Bioinformatics</em> 17 (1): 388. <a href="https://doi.org/10.1186/s12859-016-1241-0" class="external-link">https://doi.org/10.1186/s12859-016-1241-0</a>.
</div>
<div id="ref-Li2009-oc" class="csl-entry">
Li, H, and R Durbin. 2009. <span>“Fast and Accurate Short Read Alignment with <span>Burrows-Wheeler</span> Transform.”</span> <em>Bioinformatics</em> 25 (14): 1754–60. <a href="https://doi.org/10.1093/bioinformatics/btp324" class="external-link">https://doi.org/10.1093/bioinformatics/btp324</a>.
</div>
<div id="ref-Li2013-oy" class="csl-entry">
Li, Heng. 2013. <span>“Aligning Sequence Reads, Clone Sequences and Assembly Contigs with <span>BWA-MEM</span>.”</span> <em>arXiv [q-Bio.GN]</em>, March. <a href="http://arxiv.org/abs/1303.3997" class="external-link">http://arxiv.org/abs/1303.3997</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="http://girke.bioinformatics.ucr.edu/" class="external-link">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
