<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>ChIP-Seq Workflow Template • systemPipeRdata</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="ChIP-Seq Workflow Template">
<meta property="og:description" content="systemPipeRdata">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeRdata</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/systemPipeRdata.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/systemPipeChIPseq.html">ChIP-Seq Workflow Template</a>
    </li>
    <li>
      <a href="../articles/systemPipeRIBOseq.html">RIBO-Seq Workflow Template</a>
    </li>
    <li>
      <a href="../articles/systemPipeRNAseq.html">RNA-Seq Workflow Template</a>
    </li>
    <li>
      <a href="../articles/systemPipeVARseq.html">VAR-Seq Workflow Template</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/tgirke/systemPipeRdata/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>ChIP-Seq Workflow Template</h1>
                        <h4 data-toc-skip class="author">Author: Daniela Cassol (<a href="mailto:danielac@ucr.edu" class="email">danielac@ucr.edu</a>) and Thomas Girke (<a href="mailto:thomas.girke@ucr.edu" class="email">thomas.girke@ucr.edu</a>)</h4>
            
            <h4 data-toc-skip class="date">Last update: 27 April, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/tgirke/systemPipeRdata/blob/HEAD/vignettes/systemPipeChIPseq.Rmd" class="external-link"><code>vignettes/systemPipeChIPseq.Rmd</code></a></small>
      <div class="hidden name"><code>systemPipeChIPseq.Rmd</code></div>

    </div>

    
    
<!--
# Compile from command-line
Rscript -e "rmarkdown::render('systemPipeChIPseq.Rmd', c('BiocStyle::html_document'), clean=FALSE)"
-->
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Users want to provide here background information about the design of their ChIP-Seq project.</p>
<div class="section level3">
<h3 id="background-and-objectives">Background and objectives<a class="anchor" aria-label="anchor" href="#background-and-objectives"></a>
</h3>
<p>This report describes the analysis of several ChIP-Seq experiments studying the DNA binding patterns of the transcriptions factors … from <em>organism</em> ….</p>
</div>
<div class="section level3">
<h3 id="experimental-design">Experimental design<a class="anchor" aria-label="anchor" href="#experimental-design"></a>
</h3>
<p>Typically, users want to specify here all information relevant for the analysis of their NGS study. This includes detailed descriptions of FASTQ files, experimental design, reference genome, gene annotations, etc.</p>
</div>
</div>
<div class="section level2">
<h2 id="samples-and-environment-settings">Samples and environment settings<a class="anchor" aria-label="anchor" href="#samples-and-environment-settings"></a>
</h2>
<div class="section level3">
<h3 id="environment-settings-and-input-data">Environment settings and input data<a class="anchor" aria-label="anchor" href="#environment-settings-and-input-data"></a>
</h3>
<p><a href="http://bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html" class="external-link"><em>systemPipeRdata</em></a> package is a helper package to generate a fully populated <a href="http://bioconductor.org/packages/release/bioc/html/systemPipeR.html" class="external-link"><em>systemPipeR</em></a> workflow environment in the current working directory with a single command. All the instruction for generating the workflow are provide in the <em>systemPipeRdata</em> vignette <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRdata.html#1_Introduction" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">systemPipeRdata</span><span class="fu">::</span><span class="fu"><a href="../reference/genWorkenvir.html">genWorkenvir</a></span><span class="op">(</span>workflow <span class="op">=</span> <span class="st">"chipseq"</span>, mydirname <span class="op">=</span> <span class="st">"chipseq"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"chipseq"</span><span class="op">)</span></code></pre></div>
<p>After building and loading the workflow environment generated by <code>genWorkenvir</code> from <em>systemPipeRdata</em> all data inputs are stored in a <code>data/</code> directory and all analysis results will be written to a separate <code>results/</code> directory, while the <code>systemPipeChIPseq.Rmd</code> script and the <code>targets</code> file are expected to be located in the parent directory. The R session is expected to run from this parent directory. Additional parameter files are stored under <code>param/</code>.</p>
<p>The chosen data set used by this report <a href="http://www.ncbi.nlm.nih.gov/sra/?term=SRP010938" class="external-link">SRP010938</a> contains 18 paired-end (PE) read sets from <em>Arabidposis thaliana</em> <span class="citation">(Howard et al. 2013)</span>. To minimize processing time during testing, each FASTQ file has been subsetted to 90,000-100,000 randomly sampled PE reads that map to the first 100,000 nucleotides of each chromosome of the <em>A. thaliana</em> genome. The corresponding reference genome sequence (FASTA) and its GFF annotation files have been truncated accordingly. This way the entire test sample data set is less than 200MB in storage space. A PE read set has been chosen for this test data set for flexibility, because it can be used for testing both types of analysis routines requiring either SE (single end) reads or PE reads.</p>
<p>To work with real data, users want to organize their own data similarly and substitute all test data for their own data. To rerun an established workflow on new data, the initial <code>targets</code> file along with the corresponding FASTQ files are usually the only inputs the user needs to provide.</p>
<p>For more details, please consult the documentation <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html" class="external-link">here</a>. More information about the <code>targets</code> files from <em>systemPipeR</em> can be found <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#42_Structure_of_initial_targets_data" class="external-link">here</a>.</p>
<div class="section level4">
<h4 id="experiment-definition-provided-by-targets-file">Experiment definition provided by <code>targets</code> file<a class="anchor" aria-label="anchor" href="#experiment-definition-provided-by-targets-file"></a>
</h4>
<p>The <code>targets</code> file defines all FASTQ files and sample comparisons of the analysis workflow.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">targetspath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"targetsPE_chip.txt"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>
<span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">targetspath</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span>
<span class="va">targets</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">6</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##                     FileName1                   FileName2</span>
<span class="co">## 1 ./data/SRR446027_1.fastq.gz ./data/SRR446027_2.fastq.gz</span>
<span class="co">## 2 ./data/SRR446028_1.fastq.gz ./data/SRR446028_2.fastq.gz</span>
<span class="co">## 3 ./data/SRR446029_1.fastq.gz ./data/SRR446029_2.fastq.gz</span>
<span class="co">## 4 ./data/SRR446030_1.fastq.gz ./data/SRR446030_2.fastq.gz</span>
<span class="co">##   SampleName Factor        Date SampleReference</span>
<span class="co">## 1        M1A     M1 23-Mar-2012                </span>
<span class="co">## 2        M1B     M1 23-Mar-2012                </span>
<span class="co">## 3        A1A     A1 23-Mar-2012             M1A</span>
<span class="co">## 4        A1B     A1 23-Mar-2012             M1B</span></code></pre>
<p>To work with custom data, users need to generate a <em><code>targets</code></em> file containing the paths to their own FASTQ files.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="workflow-environment">Workflow environment<a class="anchor" aria-label="anchor" href="#workflow-environment"></a>
</h2>
<p><em><code>systemPipeR</code></em> workflows can be designed and built from start to finish with a single command, importing from an R Markdown file or stepwise in interactive mode from the R console.</p>
<p>This tutorial will demonstrate how to build the workflow in an interactive mode, appending each step. The workflow is constructed by connecting each step via <code>appendStep</code> method. Each <code>SYSargsList</code> instance contains instructions needed for processing a set of input files with a specific command-line or R software and the paths to the corresponding outfiles generated by a particular tool/step.</p>
<p>To create a Workflow within <em><code>systemPipeR</code></em>, we can start by defining an empty container and checking the directory structure:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://systempipe.org/" class="external-link">systemPipeR</a></span><span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SPRproject.html" class="external-link">SPRproject</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">sal</span></code></pre></div>
<div class="section level3">
<h3 id="required-packages-and-resources">Required packages and resources<a class="anchor" aria-label="anchor" href="#required-packages-and-resources"></a>
</h3>
<p>The <code>systemPipeR</code> package needs to be loaded <span class="citation">(H Backman and Girke 2016)</span>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://systempipe.org/" class="external-link">systemPipeR</a></span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"load_SPR"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="read-preprocessing">Read preprocessing<a class="anchor" aria-label="anchor" href="#read-preprocessing"></a>
</h3>
<div class="section level4">
<h4 id="fastq-quality-report">FASTQ quality report<a class="anchor" aria-label="anchor" href="#fastq-quality-report"></a>
</h4>
<p>The following <code>seeFastq</code> and <code>seeFastqPlot</code> functions generate and plot a series of useful quality statistics for a set of FASTQ files, including per cycle quality box plots, base proportions, base-level quality trends, relative k-mer diversity, length, and occurrence distribution of reads, number of reads above quality cutoffs and mean quality distribution. The results are written to a PDF file named <code>fastqReport.pdf</code>.</p>
<p>This is the pre-trimming fastq report. Another post-trimming fastq report step is not included in the default. It is recommended to run this step first to decide whether the trimming is needed.</p>
<p>Please note that initial targets files are being used here. In this case, it has been added to the first step using the <code>updateColumn</code> function, and later, we used the <code>getColumn</code> function to extract a named vector.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"targetsPE_chip.txt"</span>, comment.char <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span>
    <span class="fu">updateColumn</span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"load_SPR"</span>, position <span class="op">=</span> <span class="st">"targetsWF"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">targets</span>
    <span class="va">fq_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"load_SPR"</span>, <span class="st">"targetsWF"</span>, column <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
    <span class="va">fqlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html" class="external-link">seeFastq</a></span><span class="op">(</span>fastq <span class="op">=</span> <span class="va">fq_files</span>, batchsize <span class="op">=</span> <span class="fl">10000</span>, klength <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span><span class="st">"./results/fastqReport.pdf"</span>, height <span class="op">=</span> <span class="fl">18</span>, width <span class="op">=</span> <span class="fl">4</span> <span class="op">*</span>
        <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">fqlist</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/seeFastq.html" class="external-link">seeFastqPlot</a></span><span class="op">(</span><span class="va">fqlist</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"fastq_report"</span>, dependency <span class="op">=</span> <span class="st">"load_SPR"</span><span class="op">)</span></code></pre></div>
<img src="results/fastqReport.png"><div align="center">
Figure 1: FASTQ quality report for 18 samples
</div>
<p></p>
</div>
<div class="section level4">
<h4 id="preprocessing-with-preprocessreads-function">Preprocessing with <code>preprocessReads</code> function<a class="anchor" aria-label="anchor" href="#preprocessing-with-preprocessreads-function"></a>
</h4>
<p>The function <code>preprocessReads</code> allows to apply predefined or custom read preprocessing functions to all FASTQ files referenced in a <code>SYSargsList</code> container, such as quality filtering or adapter trimming routines. Internally, <code>preprocessReads</code> uses the <code>FastqStreamer</code> function from the <code>ShortRead</code> package to stream through large FASTQ files in a memory-efficient manner. The following example performs adapter trimming with the <code>trimLRPatterns</code> function from the <code>Biostrings</code> package.</p>
<p>Here, we are appending this step to the <code>SYSargsList</code> object created previously. All the parameters are defined on the <code>preprocessReads-pe.yml</code> file.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"preprocessing"</span>, targets <span class="op">=</span> <span class="st">"targetsPE_chip.txt"</span>,
    dir <span class="op">=</span> <span class="cn">TRUE</span>, wf_file <span class="op">=</span> <span class="st">"preprocessReads/preprocessReads-pe.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"preprocessReads/preprocessReads-pe.yml"</span>, dir_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl"</span>,
        package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>,
        FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>,
    dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fastq_report"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>After the preprocessing step, the <code>outfiles</code> files can be used to generate the new targets files containing the paths to the trimmed FASTQ files. The new targets information can be used for the next workflow step instance, <em>e.g.</em> running the NGS alignments with the trimmed FASTQ files. The <code>appendStep</code> function is automatically handling this connectivity between steps. Please check the next step for more details.</p>
<p>The following example shows how one can design a custom read <em>‘preprocessReads’</em> function using utilities provided by the <code>ShortRead</code> package, and then run it in batch mode with the <em>‘preprocessReads’</em> function. Here, it is possible to replace the function used on the <code>preprocessing</code> step and modify the <code>sal</code> object. Because it is a custom function, it is necessary to save the part in the R object, and internally the <code>preprocessReads.doc.R</code> is loading the custom function. If the R object is saved with a different name (here <code>"param/customFCT.RData"</code>), please replace that accordingly in the <code>preprocessReads.doc.R</code>.</p>
<p>Please, note that this step is not added to the workflow, here just for demonstration.</p>
<p>First, we defined the custom function in the workflow:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">filterFct</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">fq</span>, <span class="va">cutoff</span> <span class="op">=</span> <span class="fl">20</span>, <span class="va">Nexceptions</span> <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span>
        <span class="va">qcount</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/QualityScaledXStringSet-class.html" class="external-link">quality</a></span><span class="op">(</span><span class="va">fq</span><span class="op">)</span>, <span class="st">"matrix"</span><span class="op">)</span> <span class="op">&lt;=</span> <span class="va">cutoff</span>,
            na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
        <span class="co"># Retains reads where Phred scores are &gt;= cutoff</span>
        <span class="co"># with N exceptions</span>
        <span class="va">fq</span><span class="op">[</span><span class="va">qcount</span> <span class="op">&lt;=</span> <span class="va">Nexceptions</span><span class="op">]</span>
    <span class="op">}</span>
    <span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html" class="external-link">ls</a></span><span class="op">(</span><span class="op">)</span>, file <span class="op">=</span> <span class="st">"param/customFCT.RData"</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"custom_preprocessing_function"</span>, dependency <span class="op">=</span> <span class="st">"preprocessing"</span><span class="op">)</span></code></pre></div>
<p>After, we can edit the input parameter:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">yamlinput</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"preprocessing"</span><span class="op">)</span><span class="op">$</span><span class="va">Fct</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">yamlinput</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"preprocessing"</span>, <span class="st">"Fct"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"'filterFct(fq, cutoff=20, Nexceptions=0)'"</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">yamlinput</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"preprocessing"</span><span class="op">)</span><span class="op">$</span><span class="va">Fct</span>  <span class="co">## check the new function</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">cmdlist</a></span><span class="op">(</span><span class="va">sal</span>, <span class="st">"preprocessing"</span>, targets <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>  <span class="co">## check if the command line was updated with success</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="alignments">Alignments<a class="anchor" aria-label="anchor" href="#alignments"></a>
</h3>
<div class="section level4">
<h4 id="read-mapping-with-bowtie2">Read mapping with <code>Bowtie2</code><a class="anchor" aria-label="anchor" href="#read-mapping-with-bowtie2"></a>
</h4>
<p>The NGS reads of this project will be aligned with <code>Bowtie2</code> against the reference genome sequence <span class="citation">(Langmead and Salzberg 2012)</span>. The parameter settings of the Bowtie2 index are defined in the <code>bowtie2-index.cwl</code> and <code>bowtie2-index.yml</code> files.</p>
<p>Building the index:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"bowtie2_index"</span>, dir <span class="op">=</span> <span class="cn">FALSE</span>,
    targets <span class="op">=</span> <span class="cn">NULL</span>, wf_file <span class="op">=</span> <span class="st">"bowtie2/bowtie2-index.cwl"</span>, input_file <span class="op">=</span> <span class="st">"bowtie2/bowtie2-index.yml"</span>,
    dir_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>,
    inputvars <span class="op">=</span> <span class="cn">NULL</span>, dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"preprocessing"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The parameter settings of the aligner are defined in the <code>workflow_bowtie2-pe.cwl</code> and <code>workflow_bowtie2-pe.yml</code> files. The following shows how to construct the corresponding <em>SYSargsList</em> object.</p>
<p>In ChIP-Seq experiments it is usually more appropriate to eliminate reads mapping to multiple locations. To achieve this, users want to remove the argument setting <code>-k 50 non-deterministic</code> in the configuration files.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"bowtie2_alignment"</span>,
    dir <span class="op">=</span> <span class="cn">TRUE</span>, targets <span class="op">=</span> <span class="st">"targetsPE_chip.txt"</span>, wf_file <span class="op">=</span> <span class="st">"workflow-bowtie2/workflow_bowtie2-pe.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"workflow-bowtie2/workflow_bowtie2-pe.yml"</span>,
    dir_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl"</span>, package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>,
    inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>, FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>,
        SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>, dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bowtie2_index"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>To double-check the command line for each sample, please use the following:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">cmdlist</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"bowtie2_alignment"</span>, targets <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="read-and-alignment-stats">Read and alignment stats<a class="anchor" aria-label="anchor" href="#read-and-alignment-stats"></a>
</h4>
<p>The following provides an overview of the number of reads in each sample and how many of them aligned to the reference.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">fqpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"bowtie2_alignment"</span>, <span class="st">"targetsWF"</span>,
        column <span class="op">=</span> <span class="st">"FileName1"</span><span class="op">)</span>
    <span class="va">bampaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"bowtie2_alignment"</span>, <span class="st">"outfiles"</span>,
        column <span class="op">=</span> <span class="st">"samtools_sort_bam"</span><span class="op">)</span>
    <span class="va">read_statsDF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/alignStats.html" class="external-link">alignStats</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">bampaths</span>, fqpaths <span class="op">=</span> <span class="va">fqpaths</span>,
        pairEnd <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">read_statsDF</span>, <span class="st">"results/alignStats.xls"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>,
        quote <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"align_stats"</span>, dependency <span class="op">=</span> <span class="st">"bowtie2_alignment"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="create-symbolic-links-for-viewing-bam-files-in-igv">Create symbolic links for viewing BAM files in IGV<a class="anchor" aria-label="anchor" href="#create-symbolic-links-for-viewing-bam-files-in-igv"></a>
</h4>
<p>The <code>symLink2bam</code> function creates symbolic links to view the BAM alignment files in a genome browser such as IGV without moving these large files to a local system. The corresponding URLs are written to a file with a path specified under <code>urlfile</code>, here <code>IGVurl.txt</code>. Please replace the directory and the user name.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">bampaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"bowtie2_alignment"</span>, <span class="st">"outfiles"</span>,
        column <span class="op">=</span> <span class="st">"samtools_sort_bam"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/symLink2bam.html" class="external-link">symLink2bam</a></span><span class="op">(</span>sysargs <span class="op">=</span> <span class="va">bampaths</span>, htmldir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/.html/"</span>, <span class="st">"somedir/"</span><span class="op">)</span>,
        urlbase <span class="op">=</span> <span class="st">"http://cluster.hpcc.ucr.edu/~tgirke/"</span>, urlfile <span class="op">=</span> <span class="st">"./results/IGVurl.txt"</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"bam_IGV"</span>, dependency <span class="op">=</span> <span class="st">"bowtie2_alignment"</span>, run_step <span class="op">=</span> <span class="st">"optional"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="utilities-for-coverage-data">Utilities for coverage data<a class="anchor" aria-label="anchor" href="#utilities-for-coverage-data"></a>
</h3>
<p>The following introduces several utilities useful for ChIP-Seq data. They are not part of the actual workflow. These utilities can be explored once the workflow is executed.</p>
<div class="section level4">
<h4 id="rle-object-stores-coverage-information">Rle object stores coverage information<a class="anchor" aria-label="anchor" href="#rle-object-stores-coverage-information"></a>
</h4>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bampaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"bowtie2_alignment"</span>, <span class="st">"outfiles"</span>,
    column <span class="op">=</span> <span class="st">"samtools_sort_bam"</span><span class="op">)</span>
<span class="va">aligns</span> <span class="op">&lt;-</span> <span class="fu">readGAlignments</span><span class="op">(</span><span class="va">bampaths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="va">cov</span> <span class="op">&lt;-</span> <span class="fu">coverage</span><span class="op">(</span><span class="va">aligns</span><span class="op">)</span>
<span class="va">cov</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="resizing-aligned-reads">Resizing aligned reads<a class="anchor" aria-label="anchor" href="#resizing-aligned-reads"></a>
</h4>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">trim</span><span class="op">(</span><span class="fu">resize</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">aligns</span>, <span class="st">"GRanges"</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">200</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="naive-peak-calling">Naive peak calling<a class="anchor" aria-label="anchor" href="#naive-peak-calling"></a>
</h4>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">islands</span> <span class="op">&lt;-</span> <span class="fu">slice</span><span class="op">(</span><span class="va">cov</span>, lower <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>
<span class="va">islands</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="plot-coverage-for-defined-region">Plot coverage for defined region<a class="anchor" aria-label="anchor" href="#plot-coverage-for-defined-region"></a>
</h4>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://lawremi.github.io/ggbio/" class="external-link">ggbio</a></span><span class="op">)</span>
<span class="va">myloc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Chr1"</span>, <span class="fl">1</span>, <span class="fl">1e+05</span><span class="op">)</span>
<span class="va">ga</span> <span class="op">&lt;-</span> <span class="fu">readGAlignments</span><span class="op">(</span><span class="va">bampaths</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, use.names <span class="op">=</span> <span class="cn">TRUE</span>, param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html" class="external-link">ScanBamParam</a></span><span class="op">(</span>which <span class="op">=</span> <span class="fu">GRanges</span><span class="op">(</span><span class="va">myloc</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,
    <span class="fu">IRanges</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">myloc</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">myloc</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/ggbio/man/autoplot-method.html" class="external-link">autoplot</a></span><span class="op">(</span><span class="va">ga</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">strand</span>, fill <span class="op">=</span> <span class="va">strand</span><span class="op">)</span>, facets <span class="op">=</span> <span class="va">strand</span> <span class="op">~</span>
    <span class="va">seqnames</span>, stat <span class="op">=</span> <span class="st">"coverage"</span><span class="op">)</span></code></pre></div>
<p><img src="results/coverage.png"></p>
<div align="center">
Figure 2: Plot coverage for chromosome 1 region.
</div>
<p></p>
</div>
</div>
<div class="section level3">
<h3 id="peak-calling-with-macs2">Peak calling with MACS2<a class="anchor" aria-label="anchor" href="#peak-calling-with-macs2"></a>
</h3>
<div class="section level4">
<h4 id="merge-bam-files-of-replicates-prior-to-peak-calling">Merge BAM files of replicates prior to peak calling<a class="anchor" aria-label="anchor" href="#merge-bam-files-of-replicates-prior-to-peak-calling"></a>
</h4>
<p>Merging BAM files of technical and/or biological replicates can improve the sensitivity of the peak calling by increasing the depth of read coverage. The <code>mergeBamByFactor</code> function merges BAM files based on grouping information specified by a <code>factor</code>, here the <code>Factor</code> column of the imported targets file. It also returns an updated <code>targets</code> object containing the paths to the merged BAM files as well as to any unmerged files without replicates. The updated <code>targets</code> object can be used to update the <code>SYSargsList</code> object.</p>
<p>This step can be skipped if merging of BAM files is not desired.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">bampaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"bowtie2_alignment"</span>, <span class="st">"outfiles"</span>,
        column <span class="op">=</span> <span class="st">"samtools_sort_bam"</span><span class="op">)</span>
    <span class="va">merge_bams</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/mergeBamByFactor.html" class="external-link">mergeBamByFactor</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">bampaths</span>, targetsDF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">targetsWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span><span class="op">[[</span><span class="st">"bowtie2_alignment"</span><span class="op">]</span><span class="op">]</span>,
        overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
    <span class="fu">updateColumn</span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"merge_bams"</span>, position <span class="op">=</span> <span class="st">"targetsWF"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">merge_bams</span>
    <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargets.html" class="external-link">writeTargets</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"merge_bams"</span>, file <span class="op">=</span> <span class="st">"targets_merge_bams.txt"</span>,
        overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"merge_bams"</span>, dependency <span class="op">=</span> <span class="st">"bowtie2_alignment"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="peak-calling-without-inputreference-sample">Peak calling without input/reference sample<a class="anchor" aria-label="anchor" href="#peak-calling-without-inputreference-sample"></a>
</h4>
<p>MACS2 can perform peak calling on ChIP-Seq data with and without input samples <span class="citation">(Zhang et al. 2008)</span>. The following performs peak calling without input on all samples specified in the corresponding <code>targets</code> object. Note, due to the small size of the sample data, MACS2 needs to be run here with the <code>nomodel</code> setting. For real data sets, users want to remove this parameter in the corresponding <code>*.param</code> file(s).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span>,
    targets <span class="op">=</span> <span class="st">"targets_merge_bams.txt"</span>, wf_file <span class="op">=</span> <span class="st">"MACS2/macs2-noinput.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"MACS2/macs2-noinput.yml"</span>, dir_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl"</span>,
        package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>FileName <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>,
        SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>, dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"merge_bams"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="peak-calling-with-inputreference-sample">Peak calling with input/reference sample<a class="anchor" aria-label="anchor" href="#peak-calling-with-inputreference-sample"></a>
</h4>
<p>To perform peak calling with input samples, they can be most conveniently specified in the <code>SampleReference</code> column of the initial <code>targets</code> file. The <code>writeTargetsRef</code> function uses this information to create a <code>targets</code> file intermediate for running MACS2 with the corresponding input samples.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/writeTargetsRef.html" class="external-link">writeTargetsRef</a></span><span class="op">(</span>infile <span class="op">=</span> <span class="st">"targets_merge_bams.txt"</span>, outfile <span class="op">=</span> <span class="st">"targets_bam_ref.txt"</span>,
        silent <span class="op">=</span> <span class="cn">FALSE</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"writeTargetsRef"</span>, dependency <span class="op">=</span> <span class="st">"merge_bams"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList.html" class="external-link">SYSargsList</a></span><span class="op">(</span>step_name <span class="op">=</span> <span class="st">"call_peaks_macs_withref"</span>,
    targets <span class="op">=</span> <span class="st">"targets_bam_ref.txt"</span>, wf_file <span class="op">=</span> <span class="st">"MACS2/macs2-input.cwl"</span>,
    input_file <span class="op">=</span> <span class="st">"MACS2/macs2-input.yml"</span>, dir_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/cwl"</span>,
        package <span class="op">=</span> <span class="st">"systemPipeR"</span><span class="op">)</span>, inputvars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>FileName1 <span class="op">=</span> <span class="st">"_FASTQ_PATH1_"</span>,
        FileName2 <span class="op">=</span> <span class="st">"_FASTQ_PATH2_"</span>, SampleName <span class="op">=</span> <span class="st">"_SampleName_"</span><span class="op">)</span>,
    dependency <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"writeTargetsRef"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The peak calling results from MACS2 are written for each sample to separate files in the <code>results/call_peaks_macs_withref</code> directory. They are named after the corresponding files with extensions used by MACS2.</p>
</div>
<div class="section level4">
<h4 id="identify-consensus-peaks">Identify consensus peaks<a class="anchor" aria-label="anchor" href="#identify-consensus-peaks"></a>
</h4>
<p>The following example shows how one can identify consensus peaks among two peak sets sharing either a minimum absolute overlap and/or minimum relative overlap using the <code>subsetByOverlaps</code> or <code>olRanges</code> functions, respectively. Note, the latter is a custom function imported below by sourcing it.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">peaks_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span>,
        <span class="st">"outfiles"</span>, column <span class="op">=</span> <span class="st">"peaks_xls"</span><span class="op">)</span>
    <span class="va">peak_M1A</span> <span class="op">&lt;-</span> <span class="va">peaks_files</span><span class="op">[</span><span class="st">"M1A"</span><span class="op">]</span>
    <span class="va">peak_M1A</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">peak_M1A</span>, comment <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>,
        <span class="st">"GRanges"</span><span class="op">)</span>
    <span class="va">peak_A1A</span> <span class="op">&lt;-</span> <span class="va">peaks_files</span><span class="op">[</span><span class="st">"A1A"</span><span class="op">]</span>
    <span class="va">peak_A1A</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">peak_A1A</span>, comment <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>,
        <span class="st">"GRanges"</span><span class="op">)</span>
    <span class="op">(</span><span class="va">myol1</span> <span class="op">&lt;-</span> <span class="fu">subsetByOverlaps</span><span class="op">(</span><span class="va">peak_M1A</span>, <span class="va">peak_A1A</span>, minoverlap <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
    <span class="co"># Returns any overlap</span>
    <span class="va">myol2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/olRanges.html" class="external-link">olRanges</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">peak_M1A</span>, subject <span class="op">=</span> <span class="va">peak_A1A</span>, output <span class="op">=</span> <span class="st">"gr"</span><span class="op">)</span>
    <span class="co"># Returns any overlap with OL length information</span>
    <span class="va">myol2</span><span class="op">[</span><span class="fu">values</span><span class="op">(</span><span class="va">myol2</span><span class="op">)</span><span class="op">[</span><span class="st">"OLpercQ"</span><span class="op">]</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">50</span><span class="op">]</span>
    <span class="co"># Returns only query peaks with a minimum overlap of</span>
    <span class="co"># 50%</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"consensus_peaks"</span>, dependency <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="annotate-peaks-with-genomic-context">Annotate peaks with genomic context<a class="anchor" aria-label="anchor" href="#annotate-peaks-with-genomic-context"></a>
</h3>
<div class="section level4">
<h4 id="annotation-with-chipseeker-package">Annotation with <code>ChIPseeker</code> package<a class="anchor" aria-label="anchor" href="#annotation-with-chipseeker-package"></a>
</h4>
<p>The following annotates the identified peaks with genomic context information using the <code>ChIPseeker</code> package <span class="citation">(Yu, Wang, and He 2015)</span>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://guangchuangyu.github.io/software/ChIPseeker" class="external-link">ChIPseeker</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">GenomicFeatures</a></span><span class="op">)</span>
    <span class="va">peaks_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span>,
        <span class="st">"outfiles"</span>, column <span class="op">=</span> <span class="st">"peaks_xls"</span><span class="op">)</span>
    <span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html" class="external-link">makeTxDbFromGFF</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/tair10.gff"</span>,
        format <span class="op">=</span> <span class="st">"gff"</span>, dataSource <span class="op">=</span> <span class="st">"TAIR"</span>, organism <span class="op">=</span> <span class="st">"Arabidopsis thaliana"</span><span class="op">)</span><span class="op">)</span>
    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">peaks_files</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
        <span class="va">peakAnno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPseeker/man/annotatePeak.html" class="external-link">annotatePeak</a></span><span class="op">(</span><span class="va">peaks_files</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, TxDb <span class="op">=</span> <span class="va">txdb</span>,
            verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
        <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">peakAnno</span><span class="op">)</span>
        <span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"./results/"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">peaks_files</span><span class="op">)</span>,
            <span class="st">"_ChIPseeker_annotated.xls"</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">outpaths</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">peaks_files</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">outpaths</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>,
            sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
    <span class="op">}</span>
    <span class="fu">updateColumn</span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"annotation_ChIPseeker"</span>, position <span class="op">=</span> <span class="st">"outfiles"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">outpaths</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"annotation_ChIPseeker"</span>, dependency <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span><span class="op">)</span></code></pre></div>
<p>The peak annotation results are written for each peak set to separate files in the <code>results/</code> directory.</p>
<p>Summary plots provided by the <code>ChIPseeker</code> package. Here applied only to one sample for demonstration purposes.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">peaks_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span>,
        <span class="st">"outfiles"</span>, column <span class="op">=</span> <span class="st">"peaks_xls"</span><span class="op">)</span>
    <span class="va">peak</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPseeker/man/readPeakFile.html" class="external-link">readPeakFile</a></span><span class="op">(</span><span class="va">peaks_files</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span><span class="st">"results/peakscoverage.pdf"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/ChIPseeker/man/covplot.html" class="external-link">covplot</a></span><span class="op">(</span><span class="va">peak</span>, weightCol <span class="op">=</span> <span class="st">"X.log10.pvalue."</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span><span class="st">"results/peaksHeatmap.pdf"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/ChIPseeker/man/peakHeatmap.html" class="external-link">peakHeatmap</a></span><span class="op">(</span><span class="va">peaks_files</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, TxDb <span class="op">=</span> <span class="va">txdb</span>, upstream <span class="op">=</span> <span class="fl">1000</span>,
        downstream <span class="op">=</span> <span class="fl">1000</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span><span class="st">"results/peaksProfile.pdf"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/ChIPseeker/man/plotAvgProf2.html" class="external-link">plotAvgProf2</a></span><span class="op">(</span><span class="va">peaks_files</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, TxDb <span class="op">=</span> <span class="va">txdb</span>, upstream <span class="op">=</span> <span class="fl">1000</span>,
        downstream <span class="op">=</span> <span class="fl">1000</span>, xlab <span class="op">=</span> <span class="st">"Genomic Region (5'-&gt;3')"</span>,
        ylab <span class="op">=</span> <span class="st">"Read Count Frequency"</span>, conf <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"ChIPseeker_plots"</span>, dependency <span class="op">=</span> <span class="st">"annotation_ChIPseeker"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="annotation-with-chippeakanno-package">Annotation with <code>ChIPpeakAnno</code> package<a class="anchor" aria-label="anchor" href="#annotation-with-chippeakanno-package"></a>
</h4>
<p>Same as in previous step but using the <code>ChIPpeakAnno</code> package <span class="citation">(Zhu et al. 2010)</span> for annotating the peaks.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ChIPpeakAnno</span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">GenomicFeatures</a></span><span class="op">)</span>
    <span class="va">peaks_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span>,
        <span class="st">"outfiles"</span>, column <span class="op">=</span> <span class="st">"peaks_xls"</span><span class="op">)</span>
    <span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html" class="external-link">makeTxDbFromGFF</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/tair10.gff"</span>,
        format <span class="op">=</span> <span class="st">"gff"</span>, dataSource <span class="op">=</span> <span class="st">"TAIR"</span>, organism <span class="op">=</span> <span class="st">"Arabidopsis thaliana"</span><span class="op">)</span><span class="op">)</span>
    <span class="va">ge</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">genes</a></span><span class="op">(</span><span class="va">txdb</span>, columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tx_name"</span>, <span class="st">"gene_id"</span>, <span class="st">"tx_type"</span><span class="op">)</span><span class="op">)</span>
    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">peaks_files</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
        <span class="va">peaksGR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">peaks_files</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, comment <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span>,
            <span class="st">"GRanges"</span><span class="op">)</span>
        <span class="va">annotatedPeak</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/annotatePeakInBatch.html" class="external-link">annotatePeakInBatch</a></span><span class="op">(</span><span class="va">peaksGR</span>, AnnotationData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">genes</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span><span class="op">)</span>
        <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">annotatedPeak</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">values</a></span><span class="op">(</span><span class="va">ge</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">values</a></span><span class="op">(</span><span class="va">annotatedPeak</span><span class="op">)</span><span class="op">$</span><span class="va">feature</span>,
            <span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
        <span class="va">df</span><span class="op">$</span><span class="va">tx_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">tx_name</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>,
            sep <span class="op">=</span> <span class="st">""</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
        <span class="va">df</span><span class="op">$</span><span class="va">tx_type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">tx_type</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>,
            sep <span class="op">=</span> <span class="st">""</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
        <span class="va">outpaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"./results/"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">peaks_files</span><span class="op">)</span>,
            <span class="st">"_ChIPpeakAnno_annotated.xls"</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">outpaths</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">peaks_files</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">df</span>, <span class="va">outpaths</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>,
            sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
    <span class="op">}</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"annotation_ChIPpeakAnno"</span>, dependency <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span>,
    run_step <span class="op">=</span> <span class="st">"optional"</span><span class="op">)</span></code></pre></div>
<p>The peak annotation results are written for each peak set to separate files in the <code>results/</code> directory.</p>
</div>
</div>
<div class="section level3">
<h3 id="count-reads-overlapping-peaks">Count reads overlapping peaks<a class="anchor" aria-label="anchor" href="#count-reads-overlapping-peaks"></a>
</h3>
<p>The <code>countRangeset</code> function is a convenience wrapper to perform read counting iteratively over several range sets, here peak range sets. Internally, the read counting is performed with the <code>summarizeOverlaps</code> function from the <code>GenomicAlignments</code> package. The resulting count tables are directly saved to files, one for each peak set.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span>
    <span class="va">bam_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"bowtie2_alignment"</span>, <span class="st">"outfiles"</span>,
        column <span class="op">=</span> <span class="st">"samtools_sort_bam"</span><span class="op">)</span>
    <span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span>, <span class="st">"outfiles"</span>,
        column <span class="op">=</span> <span class="st">"peaks_xls"</span><span class="op">)</span>
    <span class="va">outfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"./results/"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span>, <span class="st">"_countDF.xls"</span><span class="op">)</span>
    <span class="va">bfl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFileList</a></span><span class="op">(</span><span class="va">bam_files</span>, yieldSize <span class="op">=</span> <span class="fl">50000</span>, index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
    <span class="va">countDFnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/countRangeset.html" class="external-link">countRangeset</a></span><span class="op">(</span><span class="va">bfl</span>, <span class="va">args</span>, <span class="va">outfiles</span>, mode <span class="op">=</span> <span class="st">"Union"</span>,
        ignore.strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
    <span class="fu">updateColumn</span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"count_peak_ranges"</span>, position <span class="op">=</span> <span class="st">"outfiles"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">countDFnames</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"count_peak_ranges"</span>, dependency <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span>,
    <span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="differential-binding-analysis">Differential binding analysis<a class="anchor" aria-label="anchor" href="#differential-binding-analysis"></a>
</h3>
<p>The <code>runDiff</code> function performs differential binding analysis in batch mode for several count tables using <code>edgeR</code> or <code>DESeq2</code> <span class="citation">(Robinson, McCarthy, and Smyth 2010; Love, Huber, and Anders 2014)</span>. Internally, it calls the functions <code>run_edgeR</code> and <code>run_DESeq2</code>. It also returns the filtering results and plots from the downstream <code>filterDEGs</code> function using the fold change and FDR cutoffs provided under the <code>dbrfilter</code> argument.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">countDF_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"count_peak_ranges"</span>,
        <span class="st">"outfiles"</span><span class="op">)</span>
    <span class="va">outfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"./results/"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">countDF_files</span><span class="op">)</span>, <span class="st">"_peaks_edgeR.xls"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">outfiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">countDF_files</span><span class="op">)</span>
    <span class="va">cmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/readComp.html" class="external-link">readComp</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">stepsWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span><span class="op">[[</span><span class="st">"bowtie2_alignment"</span><span class="op">]</span><span class="op">]</span>,
        format <span class="op">=</span> <span class="st">"matrix"</span><span class="op">)</span>
    <span class="va">dbrlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runDiff.html" class="external-link">runDiff</a></span><span class="op">(</span>args <span class="op">=</span> <span class="va">countDF_files</span>, outfiles <span class="op">=</span> <span class="va">outfiles</span>,
        diffFct <span class="op">=</span> <span class="va">run_edgeR</span>, targets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">targetsWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span><span class="op">[[</span><span class="st">"bowtie2_alignment"</span><span class="op">]</span><span class="op">]</span>,
        cmp <span class="op">=</span> <span class="va">cmp</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, independent <span class="op">=</span> <span class="cn">TRUE</span>, dbrfilter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span>Fold <span class="op">=</span> <span class="fl">2</span>,
            FDR <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"diff_bind_analysis"</span>, dependency <span class="op">=</span> <span class="st">"count_peak_ranges"</span>,
    <span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="go-term-enrichment-analysis">GO term enrichment analysis<a class="anchor" aria-label="anchor" href="#go-term-enrichment-analysis"></a>
</h3>
<p>The following performs GO term enrichment analysis for each annotated peak set.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="va">annofiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"annotation_ChIPseeker"</span>,
        <span class="st">"outfiles"</span><span class="op">)</span>
    <span class="va">gene_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">annofiles</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span>,
        <span class="st">"geneId"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"data/GO/catdb.RData"</span><span class="op">)</span>
    <span class="va">BatchResult</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/GOHyperGAll.html" class="external-link">GOCluster_Report</a></span><span class="op">(</span>catdb <span class="op">=</span> <span class="va">catdb</span>, setlist <span class="op">=</span> <span class="va">gene_ids</span>,
        method <span class="op">=</span> <span class="st">"all"</span>, id_type <span class="op">=</span> <span class="st">"gene"</span>, CLSZ <span class="op">=</span> <span class="fl">2</span>, cutoff <span class="op">=</span> <span class="fl">0.9</span>,
        gocats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MF"</span>, <span class="st">"BP"</span>, <span class="st">"CC"</span><span class="op">)</span>, recordSpecGO <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">BatchResult</span>, <span class="st">"results/GOBatchAll.xls"</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>,
        row.names <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"go_enrich"</span>, dependency <span class="op">=</span> <span class="st">"annotation_ChIPseeker"</span>,
    <span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="motif-analysis">Motif analysis<a class="anchor" aria-label="anchor" href="#motif-analysis"></a>
</h3>
<div class="section level4">
<h4 id="parse-dna-sequences-of-peak-regions-from-genome">Parse DNA sequences of peak regions from genome<a class="anchor" aria-label="anchor" href="#parse-dna-sequences-of-peak-regions-from-genome"></a>
</h4>
<p>Enrichment analysis of known DNA binding motifs or <em>de novo</em> discovery of novel motifs requires the DNA sequences of the identified peak regions. To parse the corresponding sequences from the reference genome, the <code>getSeq</code> function from the <code>Biostrings</code> package can be used. The following example parses the sequences for each peak set and saves the results to separate FASTA files, one for each peak set. In addition, the sequences in the FASTA files are ranked (sorted) by increasing p-values as expected by some motif discovery tools, such as <code>BCRANK</code>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">seqLogo</span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BCRANK</span><span class="op">)</span>
    <span class="va">rangefiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span>,
        <span class="st">"outfiles"</span><span class="op">)</span>
    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>along <span class="op">=</span> <span class="va">rangefiles</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
        <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">rangefiles</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, comment <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span>
        <span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">df</span>, <span class="st">"GRanges"</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqnames</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span><span class="op">)</span>,
            <span class="st">"_"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span>, <span class="st">"-"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span><span class="op">)</span>
        <span class="va">peaks</span> <span class="op">&lt;-</span> <span class="va">peaks</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">values</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span><span class="op">$</span><span class="va">X.log10.pvalue.</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span>
        <span class="va">pseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/getSeq.html" class="external-link">getSeq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/FaFile-class.html" class="external-link">FaFile</a></span><span class="op">(</span><span class="st">"./data/tair10.fasta"</span><span class="op">)</span>, <span class="va">peaks</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">pseq</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">peaks</span><span class="op">)</span>
        <span class="fu"><a href="https://rdrr.io/pkg/Biostrings/man/XStringSet-io.html" class="external-link">writeXStringSet</a></span><span class="op">(</span><span class="va">pseq</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">rangefiles</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">".fasta"</span><span class="op">)</span><span class="op">)</span>
    <span class="op">}</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"parse_peak_sequences"</span>, dependency <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span>,
    <span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="motif-discovery-with-bcrank">Motif discovery with <code>BCRANK</code><a class="anchor" aria-label="anchor" href="#motif-discovery-with-bcrank"></a>
</h4>
<p>The Bioconductor package <code>BCRANK</code> is one of the many tools available for <em>de novo</em> discovery of DNA binding motifs in peak regions of ChIP-Seq experiments. The given example applies this method on the first peak sample set and plots the sequence logo of the highest ranking motif.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">seqLogo</span><span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BCRANK</span><span class="op">)</span>
    <span class="va">rangefiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">getColumn</a></span><span class="op">(</span><span class="va">sal</span>, step <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span>,
        <span class="st">"outfiles"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>
    <span class="va">BCRANKout</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BCRANK/man/bcrank.html" class="external-link">bcrank</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">rangefiles</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">".fasta"</span><span class="op">)</span>, restarts <span class="op">=</span> <span class="fl">25</span>,
        use.P1 <span class="op">=</span> <span class="cn">TRUE</span>, use.P2 <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/BCRANK/man/BCRANKresult-class.html" class="external-link">toptable</a></span><span class="op">(</span><span class="va">BCRANKout</span><span class="op">)</span>
    <span class="va">topMotif</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BCRANK/man/BCRANKresult-class.html" class="external-link">toptable</a></span><span class="op">(</span><span class="va">BCRANKout</span>, <span class="fl">1</span><span class="op">)</span>
    <span class="va">weightMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/seqLogo/man/pwm-class.html" class="external-link">pwm</a></span><span class="op">(</span><span class="va">topMotif</span>, normalize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
    <span class="va">weightMatrixNormalized</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/seqLogo/man/pwm-class.html" class="external-link">pwm</a></span><span class="op">(</span><span class="va">topMotif</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span><span class="op">(</span><span class="st">"results/seqlogo.pdf"</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/pkg/seqLogo/man/seqLogo.html" class="external-link">seqLogo</a></span><span class="op">(</span><span class="va">weightMatrixNormalized</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"bcrank_enrich"</span>, dependency <span class="op">=</span> <span class="st">"call_peaks_macs_noref"</span>,
    <span class="op">)</span></code></pre></div>
<p><img src="results/seqlogo.png"></p>
<div align="center">
Figure 3: One of the motifs identified by <code>BCRANK</code>
</div>
<p></p>
</div>
</div>
<div class="section level3">
<h3 id="version-information">Version Information<a class="anchor" aria-label="anchor" href="#version-information"></a>
</h3>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">appendStep</span><span class="op">(</span><span class="va">sal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/LineWise-class.html" class="external-link">LineWise</a></span><span class="op">(</span>code <span class="op">=</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="op">}</span>, step_name <span class="op">=</span> <span class="st">"sessionInfo"</span>, dependency <span class="op">=</span> <span class="st">"bcrank_enrich"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-workflow">Running workflow<a class="anchor" aria-label="anchor" href="#running-workflow"></a>
</h2>
<div class="section level3">
<h3 id="interactive-job-submissions-in-a-single-machine">Interactive job submissions in a single machine<a class="anchor" aria-label="anchor" href="#interactive-job-submissions-in-a-single-machine"></a>
</h3>
<p>For running the workflow, <code>runWF</code> function will execute all the steps store in the workflow container. The execution will be on a single machine without submitting to a queuing system of a computer cluster.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html" class="external-link">runWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="parallelization-on-clusters">Parallelization on clusters<a class="anchor" aria-label="anchor" href="#parallelization-on-clusters"></a>
</h3>
<p>Alternatively, the computation can be greatly accelerated by processing many files in parallel using several compute nodes of a cluster, where a scheduling/queuing system is used for load balancing.</p>
<p>The <code>resources</code> list object provides the number of independent parallel cluster processes defined under the <code>Njobs</code> element in the list. The following example will run 18 processes in parallel using each 4 CPU cores. If the resources available on a cluster allow running all 18 processes at the same time, then the shown sample submission will utilize in a total of 72 CPU cores.</p>
<p>Note, <code>runWF</code> can be used with most queueing systems as it is based on utilities from the <code>batchtools</code> package, which supports the use of template files (<em><code>*.tmpl</code></em>) for defining the run parameters of different schedulers. To run the following code, one needs to have both a <code>conffile</code> (see <em><code>.batchtools.conf.R</code></em> samples <a href="https://mllg.github.io/batchtools/" class="external-link">here</a>) and a <code>template</code> file (see <em><code>*.tmpl</code></em> samples <a href="https://github.com/mllg/batchtools/tree/master/inst/templates" class="external-link">here</a>) for the queueing available on a system. The following example uses the sample <code>conffile</code> and <code>template</code> files for the Slurm scheduler provided by this package.</p>
<p>The resources can be appended when the step is generated, or it is possible to add these resources later, as the following example using the <code>addResources</code> function:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">resources</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>conffile<span class="op">=</span><span class="st">".batchtools.conf.R"</span>,
                  template<span class="op">=</span><span class="st">"batchtools.slurm.tmpl"</span>, 
                  Njobs<span class="op">=</span><span class="fl">18</span>, 
                  walltime<span class="op">=</span><span class="fl">120</span>, <span class="co">## minutes</span>
                  ntasks<span class="op">=</span><span class="fl">1</span>,
                  ncpus<span class="op">=</span><span class="fl">4</span>, 
                  memory<span class="op">=</span><span class="fl">1024</span>, <span class="co">## Mb</span>
                  partition <span class="op">=</span> <span class="st">"short"</span>
                  <span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">addResources</a></span><span class="op">(</span><span class="va">sal</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bowtie2_alignment"</span><span class="op">)</span>, resources <span class="op">=</span> <span class="va">resources</span><span class="op">)</span>
<span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/runWF.html" class="external-link">runWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-workflow">Visualize workflow<a class="anchor" aria-label="anchor" href="#visualize-workflow"></a>
</h3>
<p><em><code>systemPipeR</code></em> workflows instances can be visualized with the <code>plotWF</code> function.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/plotWF.html" class="external-link">plotWF</a></span><span class="op">(</span><span class="va">sal</span>, rstudio <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="checking-workflow-status">Checking workflow status<a class="anchor" aria-label="anchor" href="#checking-workflow-status"></a>
</h3>
<p>To check the summary of the workflow, we can use:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span>
<span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/SYSargsList-class.html" class="external-link">statusWF</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="accessing-logs-report">Accessing logs report<a class="anchor" aria-label="anchor" href="#accessing-logs-report"></a>
</h3>
<p><em><code>systemPipeR</code></em> compiles all the workflow execution logs in one central location, making it easier to check any standard output (<code>stdout</code>) or standard error (<code>stderr</code>) for any command-line tools used on the workflow or the R code stdout.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/systemPipeR/man/renderLogs.html" class="external-link">renderLogs</a></span><span class="op">(</span><span class="va">sal</span><span class="op">)</span></code></pre></div>
<p>If you are running on a single machine, use following code as an example to check if some tools used in this workflow are in your environment <strong>PATH</strong>. No warning message should be shown if all tools are installed.</p>
</div>
</div>
<div class="section level2">
<h2 id="funding">Funding<a class="anchor" aria-label="anchor" href="#funding"></a>
</h2>
<p>This project was supported by funds from the National Institutes of Health (NIH) and the National Science Foundation (NSF).</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-H_Backman2016-bt" class="csl-entry">
H Backman, Tyler W, and Thomas Girke. 2016. <span>“<span class="nocase">systemPipeR: NGS workflow and report generation environment</span>.”</span> <em>BMC Bioinformatics</em> 17 (1): 388. <a href="https://doi.org/10.1186/s12859-016-1241-0" class="external-link">https://doi.org/10.1186/s12859-016-1241-0</a>.
</div>
<div id="ref-Howard2013-fq" class="csl-entry">
Howard, Brian E, Qiwen Hu, Ahmet Can Babaoglu, Manan Chandra, Monica Borghi, Xiaoping Tan, Luyan He, et al. 2013. <span>“High-Throughput <span>RNA</span> Sequencing of Pseudomonas-Infected Arabidopsis Reveals Hidden Transcriptome Complexity and Novel Splice Variants.”</span> <em>PLoS One</em> 8 (10): e74183. <a href="https://doi.org/10.1371/journal.pone.0074183" class="external-link">https://doi.org/10.1371/journal.pone.0074183</a>.
</div>
<div id="ref-Langmead2012-bs" class="csl-entry">
Langmead, Ben, and Steven L Salzberg. 2012. <span>“Fast Gapped-Read Alignment with Bowtie 2.”</span> <em>Nat. Methods</em> 9 (4): 357–59. <a href="https://doi.org/10.1038/nmeth.1923" class="external-link">https://doi.org/10.1038/nmeth.1923</a>.
</div>
<div id="ref-Love2014-sh" class="csl-entry">
Love, Michael, Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for <span class="nocase">RNA-seq</span> Data with <span>DESeq2</span>.”</span> <em>Genome Biol.</em> 15 (12): 550. <a href="https://doi.org/10.1186/s13059-014-0550-8" class="external-link">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-Robinson2010-uk" class="csl-entry">
Robinson, M D, D J McCarthy, and G K Smyth. 2010. <span>“edgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.”</span> <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616" class="external-link">https://doi.org/10.1093/bioinformatics/btp616</a>.
</div>
<div id="ref-Yu2015-xu" class="csl-entry">
Yu, Guangchuang, Li-Gen Wang, and Qing-Yu He. 2015. <span>“<span>ChIPseeker</span>: An <span>R/Bioconductor</span> Package for <span>ChIP</span> Peak Annotation, Comparison and Visualization.”</span> <em>Bioinformatics</em> 31 (14): 2382–83. <a href="https://doi.org/10.1093/bioinformatics/btv145" class="external-link">https://doi.org/10.1093/bioinformatics/btv145</a>.
</div>
<div id="ref-Zhang2008-pc" class="csl-entry">
Zhang, Y, T Liu, C A Meyer, J Eeckhoute, D S Johnson, B E Bernstein, C Nussbaum, et al. 2008. <span>“Model-Based Analysis of <span>ChIP-Seq</span> (<span>MACS</span>).”</span> <em>Genome Biol.</em> 9 (9). <a href="https://doi.org/10.1186/gb-2008-9-9-r137" class="external-link">https://doi.org/10.1186/gb-2008-9-9-r137</a>.
</div>
<div id="ref-Zhu2010-zo" class="csl-entry">
Zhu, Lihua J, Claude Gazin, Nathan D Lawson, Hervé Pagès, Simon M Lin, David S Lapointe, and Michael R Green. 2010. <span>“<span>ChIPpeakAnno</span>: A Bioconductor Package to Annotate <span class="nocase">ChIP-seq</span> and <span class="nocase">ChIP-chip</span> Data.”</span> <em>BMC Bioinformatics</em> 11: 237. <a href="https://doi.org/10.1186/1471-2105-11-237" class="external-link">https://doi.org/10.1186/1471-2105-11-237</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="http://girke.bioinformatics.ucr.edu/" class="external-link">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
