<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate workflow templates — genWorkenvir • systemPipeRdata</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Generate workflow templates — genWorkenvir"><meta property="og:description" content="Generates workflow templates for systemPipeR package. The template environemts 
contain a predifined directory structure along with run parameter files and sample data. 
The strucutre of the workflow templates and the sample data are described in all
details in the Overview Vignette of the systemPipeR package."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">systemPipeRdata</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.23.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/systemPipeRdata.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/systemPipeChIPseq.html">ChIP-Seq Workflow Template</a>
    </li>
    <li>
      <a href="../articles/systemPipeRIBOseq.html">RIBO-Seq Workflow Template</a>
    </li>
    <li>
      <a href="../articles/systemPipeRNAseq.html">RNA-Seq Workflow Template</a>
    </li>
    <li>
      <a href="../articles/systemPipeVARseq.html">VAR-Seq Workflow Template</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/tgirke/systemPipeRdata/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate workflow templates</h1>
    
    <div class="hidden name"><code>genWorkenvir.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generates workflow templates for <code>systemPipeR</code> package. The template environemts 
contain a predifined directory structure along with run parameter files and sample data. 
The strucutre of the workflow templates and the sample data are described in all
details in the Overview Vignette of the <a href="http://bioconductor.org/packages/systemPipeR" class="external-link">systemPipeR package</a>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">genWorkenvir</span><span class="op">(</span><span class="va">workflow</span>, mydirname<span class="op">=</span><span class="cn">NULL</span>, bam<span class="op">=</span><span class="cn">FALSE</span>, ref<span class="op">=</span><span class="st">"master"</span>, subdir<span class="op">=</span><span class="cn">NULL</span>, url<span class="op">=</span><span class="cn">NULL</span>, urlname<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>workflow</dt>
<dd><p><code>character string</code> of workflow templates to be generated. Supported values 
  can be checked with the <code>\link{availableWF}()</code> function. Workflow name 
  containing a '/' are treated as GitHub repositories and installed using 
  <code>BiocManager::\link{install_github}()</code>. For Github workflow templates, the format
   <code>systemPipeR/repo</code> is required. See Details.</p></dd>
<dt>mydirname</dt>
<dd><p>Specifies the name of the workflow directory. The default <code>NULL</code> uses the 
  name of the chosen workflow. An error is issued if a directory of the same name 
  and path exists already.</p></dd>
<dt>bam</dt>
<dd><p>If <code>bam=TRUE</code> pregenerated short read alignment (BAM) files will be included
  in the <code>results</code> directory of the workflow environment. Note, these BAM files
  have been generated with the HISAT2 aligner using the FASTQ files provided in the
  <code>data</code> directory. The default <code>bam=FALSE</code> omits this step meaning no 
  BAM files will be copied into the <code>results</code> directory.</p></dd>
<dt>ref</dt>
<dd><p>Desired GitHub reference for the <code>branch</code> name. Default to <code>master</code> branch.</p></dd>
<dt>subdir</dt>
<dd><p>subdirectory within GitHub repo that contains the R package, if it is required.</p></dd>
<dt>url</dt>
<dd><p><code>character string</code> of a specifies the <code>URL</code> of a different version of 
  the <code>R Markdown</code> workflow template or other file to download. 
  The default <code>NULL</code> copies the current version available 
  in the `systemPipeRdata` or the workflow template on GitHub being selected.</p></dd>
<dt>urlname</dt>
<dd><p><code>character string</code> with the name where the downloaded file is saved.
  This is argument is required when the <code>url</code> is provided.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Workflow directory containing sample data and parameter files along with the following subdirectories:</p>
<dl><dt>param/</dt>
<dd><p>stores parameter files</p></dd>
<dt>data/</dt>
<dd><p>stores input data</p></dd>
<dt>results/</dt>
<dd><p>stores output results</p></dd>
</dl><p>For more details, please consult the Overview Vignette (HTML) of the systemPipeR package (http://bioconductor.org/packages/systemPipeR).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>When installing GitHub Workflow Packages under <a href="https://github.com/systemPipeR" class="external-link">systemPipeR Organization</a>, 
it is requested to the remotes package function <code>BiocManager::\link{install_github}()</code>
to build vignettes and also install all the dependencies, via <code>dependencies=TRUE</code>, <code>build_vignettes=TRUE</code>.</p>
<p>Check the output of <code>\link{availableWF}()</code> to the current workflow templates
available on <code>systemPipeR</code> Organization. The argument <code></code></p>
<p>For an <code><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive()</a></code> session, the <code><a href="https://rdrr.io/r/base/readline.html" class="external-link">readline()</a></code> function provides the option 
choose between proceeding or not, through options: <code>yes</code> or <code>no</code>.
For non-interactive use, if there is no package install, the option <code>yes</code> will be selected.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Thomas Girke and Daniela Cassol</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co">## Return location of sample data</span></span>
<span class="r-in"><span class="va">samplepaths</span> <span class="op">&lt;-</span> <span class="fu"><a href="pathList.html">pathList</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"><span class="co">## Generate varseq workflow environment</span></span>
<span class="r-in"><span class="fu">genWorkenvir</span><span class="op">(</span>workflow<span class="op">=</span><span class="st">"varseq"</span>, mydirname<span class="op">=</span><span class="cn">NULL</span>, url<span class="op">=</span><span class="cn">NULL</span>, urlname<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"varseq"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## List Workflow Templates from \code{systemPipeRdata} package and \code{systemPipeR} Organization </span></span>
<span class="r-in"><span class="fu"><a href="availableWF.html">availableWF</a></span><span class="op">(</span>github <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-in"><span class="co">## Generate 'systemPipeR/systemPipeChIPseq' workflow environment </span></span>
<span class="r-in"><span class="fu">genWorkenvir</span><span class="op">(</span>workflow<span class="op">=</span><span class="st">"systemPipeR/systemPipeChIPseq"</span>, mydirname<span class="op">=</span><span class="cn">NULL</span>, ref<span class="op">=</span><span class="st">"master"</span>, subdir<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"systemPipeChIPseq"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## Download a specific R Markdown file</span></span>
<span class="r-in"><span class="fu">genWorkenvir</span><span class="op">(</span>workflow<span class="op">=</span><span class="st">"systemPipeR/systemPipeRNAseq"</span>, mydirname<span class="op">=</span><span class="st">"rnaseq"</span>, url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/systemPipeR/systemPipeRNAseq/cluster/vignettes/systemPipeRNAseq.Rmd"</span>, urlname <span class="op">=</span> <span class="st">"rnaseq_V-cluster.Rmd"</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="http://girke.bioinformatics.ucr.edu/" class="external-link">Thomas Girke</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

